---
title: "RAS signature validation CCLE: Project 002"
author: "philip.east@crick.ac.uk"
date: 'Compiled: `r format(Sys.time(), "%d %B, %Y @ %H:%M:%S")`'
output:
  html_document:
    df_print: tibble
    toc: true
    toc_depth: 5
    toc_float: true
    code_folding: hide
    fig_caption: yes
---

```{css setup_css, echo=FALSE}
body .main-container {
  max-width: 80%;
}
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE, message=FALSE, warning=FALSE)
options(scipen=15)
```

# Introduction

This script validates a set of RAS transcriptional signatures against
the LUNG CCLE set of cancer cell lines. The high performing signatures
are then used to build a meta-signautre for RAS trasncriptional
activity.

The CCLE microarray expression and mutation data were obtained from
the CCLE legacy repository hosted at [The Broad](https://data.broadinstitute.org/ccle_legacy_data )

We selected lung derived cell line data. We labelled the cell lines as either KRAS mutation positive, RAS pathway mutation positive or RAS pathway mutation negative. Cell lines were labelled as RAS pathway mutation positive if they had a mutation in a RAS pathway gene member other than KRAS. RAS pathway membership was defined in [RAS pathway ref]. We removed cell lines with mutations in RAS pathway members other than KRAS from further analysis selecting 166 cell lines. We filtered the RAS signatures for genes relevant in the context of lung cancer. To do this we modelled the log COV of expression against the mean log2 expression across all genes using loess regression. Signature genes with positive residuals with respect to the fit and a higher log2 mean expression of 6 were selected. 
We mapped each of the signatures to the cell line expression data and clustered the cell-lines using hierarchical clustering with a Ward.D2 agglomeration method. We split the resultant dendrograms into three clusters, labelling each high, medium and low for signature expression. The labels were assigned based on ranked mean cluster expression. To assess the performance of each of the signatures we calculated the significance of association of the high and low clusters with KRAS mutation using a chisq test. We also calculated the log2 fold change in the means between the high and the low group and the percentage of cell lines classified as high or low. To further refine each of the best performing signatures, we ran a differential analysis between the high and low clusters using limma. We selected signature genes with an fdr < 0.05 as those driving the clustering. We merged the differential genes for each signature tested to form our meta-signature RAS84.

```{r r_init}
library( affy )
library( limma )
library( reshape2 )
library( scales )
library( RColorBrewer )
library( ComplexHeatmap )
library( gplots )
library( gridExtra )
library( gtable )
library( ggthemes )
library( Hmisc )
library( vcd )
library( UpSetR )
library( dplyr )
library( monocle )
library( pls )
library( tibble )
library( tidyr )
library( GenomicRanges )
library( MASS )
library( ggforce )
library( circlize )
library( ggrepel )
library( tidyr )
library( tibble )
library( broom )
library( reshape2 )
library( ggplot2 )
library( openxlsx )
library( gridExtra )
library( tidyverse )

## - colours
my_grey <- "#707173"
my_red <- "#993333"
my_orange <- "#f6ad6e"
my_green <- "#7ab51d"
my_lightgreen <- "#adcf82"
my_purple <- "#bb90bd"
my_blue <- "#336699"

plot_formatter <- function() {
    theme_bw( ) +
        theme( panel.grid.major = element_blank( ),
               panel.grid.minor = element_blank( ),
               panel.border = element_blank( ),
              panel.background = element_blank( ),
              axis.line = element_line(color = "black"),
              axis.line.x = element_line(color="black", size = 0.5 ),
              axis.line.y = element_line(color="black", size = 0.5 ),
              text = element_text( size = 12 ) )
}

## Paths
ESET_CCLE_AFFY_LUNG_FILE <- file.path( "data", "objects", "eset_ccle_affy_lung.rds" )
RAS_SIG_HOME <- file.path( "data", "resources", "signatures", "ras" )
RAS_PATHWAY_DEFINITION_FILE <- file.path( "data",
                                         "downloads",
                                         "oncogenic_RAS_pathway_defintions_Sanchez-Vega_etal.txt" )
RAS84_FEATURE_MAP_FILE <- file.path( "data", "resources", "RAS84_feature_map.xlsx" )
CONTROL_SIGNATURE_HOME <- file.path( "data", "resources", "signatures", "controls" )

## figures
ONCOGENIC_RAS_SAMPLE_COUNTS_PLOT_FILE <- file.path( "figures", "ccle_oncogenic_ras_sample_counts.pdf" )
SIGNATURE_HEATMAP_FILE <- file.path( "figures", "SIG_CCLE_HEATMAP.pdf" )
RAS84_HEATMAP_FILE <- file.path( "figures", "RAS84_CCLE_large_heatmap.pdf" )
RAS_HIGHLOW_CONTINGENCY_FILE <- file.path( "figures", "ccle_ras_highlow_contingency.pdf" )
KRAS_MUTATION_LOGLIK_FILE <- file.path( "figures", "ccle_KRAS_mutation_loglik.pdf" )
RAS84_RAW_HEATMAP <- file.path( "figures", "RAS84_raw_heatmap.pdf" )
SIGNATURE_CLUSTER_VOLCANOS_FILE <- file.path( "figures", "ccle_signature_cluster_volcanos.pdf" )
RAS84_SIGNATURE_HEATMAP_FILE <- file.path( "figures", "ccle_RAS84_heatmap.pdf" )
SIGNATURE_VOLCANO_YES_RAS84_FILE <- file.path( "figures", "ccle_RAS84_signature_volcano_yes.pdf" )
KRAS_MUTATION_PVALUE_FILE <- file.path( "figures", "ccle_kras_mutation_pvalue.pdf" )
CONTROL_SIGNATURES_VOLCANO_FILE <- file.path( "figures", "control_signatures_volcano.pdf" )
```

# Data

## Expression Data

Load CCLE ExpressionSet and filter for LUNG cell lines.

```{r CCLE_init}
eset_ccle_lung <- readRDS( ESET_CCLE_AFFY_LUNG_FILE )
names( pData( eset_ccle_lung )$mutations ) <- pData( eset_ccle_lung )$sample.id
mut_ccle_lung <- pData( eset_ccle_lung )$mutations %>%
                                       as.data.frame( ) %>%
                                       rownames_to_column( var = "gene" ) %>%
                                       gather( rowname, value, -gene ) %>%
                                       spread( gene, value )
```

## Signatures

Load RAS and control gene signatures from file. The heatmap shows the degree of overlap across the different RAS gene signatures. Many of the signatures contain unique genes highlighting the variation in RAS driven transcription given the biological context.

```{r signatures}
rasSigNames <- sub( "\\.csv", "", dir( RAS_SIG_HOME ) )
signatures_raw <- lapply( rasSigNames, function( sigFile ) {
    read.table( file = file.path( RAS_SIG_HOME, paste( sigFile, "csv", sep = "." ) ),
               sep = "\t", stringsAsFactors = FALSE )[, 1 ]
} )
names( signatures_raw ) <- make.names( rasSigNames )
```
```{r signature_overlap)
ovdat <- unique( unlist( signatures_raw ) )

ov <- do.call( 'cbind',lapply( signatures_raw, function( x ) as.numeric( ovdat %in% x ) ) )

hm <- Heatmap( ov, col = c( "white", my_purple ),
              show_heatmap_legend = FALSE,
              column_split = c( 1,1,1,1,2,1 ) )
print( hm )
```

# TCGA pan-cancer RAS pathway Definition

Read in TCGA's Oncogenic-RAS pathway definition. This is used to
filter cells for oncogenic RAS signaling activity driven by pathway
members other than RAS

```{r ras_pathway}
rasPathwayTab <- read.delim( file = RAS_PATHWAY_DEFINITION_FILE, sep = "\t" )
rasGeneSets <- list( og = subset( rasPathwayTab, OG.TSG == "OG" )$Gene,
                     tsg = subset( rasPathwayTab, OG.TSG == "TSG" )$Gene )
rasGeneSets <- lapply( rasGeneSets, as.character )
```

# Oncogenic RAS signalling in cell lines

Label lung cancer cell lines as either KRAS(+), oncogenic-RAS(+) or
oncogenic-RAS(-) using the cell-line mutation data and pathway
defintion defined above. Oncogenic-RAS(+) labelled cells are removed
from further analysis. The barchart shows the cell line count per group.

```{r RAS_annotate}
## - label CCLE samples as Oncogenic RAS negative, positive or KRAS+
oncoras_ccle_lung <- mut_ccle_lung %>%
    tidyr::gather( name, value, -rowname ) %>%
    filter( name %in% unlist( rasGeneSets ) ) %>%
    group_by( rowname ) %>%
    dplyr::summarize( onco_ras = ifelse( all( value == "no" ), "oncoras_neg",
                                 ifelse( any( name == "KRAS" & value == "yes" ), "kras_mut", "oncoras_pos" ) ) )

## - add labels to CCLE expressionSet
pData( eset_ccle_lung ) <- pData( eset_ccle_lung ) %>%
    rownames_to_column( ) %>%
    left_join( oncoras_ccle_lung, by = "rowname" )
   
## - filter ccle expression set for possible oncogenic ras activity with no KRAS mutation
eset_ccle_lung_kras <- eset_ccle_lung[ , !eset_ccle_lung$onco_ras %in% "oncoras_pos" ]

gg1 <- pData( eset_ccle_lung ) %>%
    ggplot( aes( x = onco_ras, fill = onco_ras ) ) +
    geom_bar( ) +
    labs( x = "oncogenic RAS activity" ) +
    plot_formatter( ) +
    theme( axis.text.x = element_text( angle = 90, hjust = 1, vjust = 0.5 ),
          legend.position = "none" ) +
    scale_fill_manual( values = c( my_purple, my_orange, my_grey ) )
print( gg1 )

if( !file.exists( ONCOGENIC_RAS_SAMPLE_COUNTS_PLOT_FILE ) ) {
    cairo_pdf( file = ONCOGENIC_RAS_SAMPLE_COUNTS_PLOT_FILE, width = 1.3, height = 3 )
    print( gg1 )
    dev.off( )
}
```

# Clean up signatures variance/expression

Here we assess the signature genes in the context of the cell-line
expression data. We model the COV of expression verse mean expression across all genes
using loess regression. Signature genes with postive
residuals with respect to the fit and a higher mean expression of 6 across
all cells are selected. The scatter plots show COV against mean expression with the loess curve shown in black and the selected genes highlighted in green.

```{r sig_assess}
plotdat_sig_filt_l <- lapply( signatures_raw, function( sig ) {
  as.data.frame( exprs( eset_ccle_lung_kras ) ) %>%
    mutate( gsymbol = fData( eset_ccle_lung_kras )$Description ) %>%
    tidyr::gather(name, value, -gsymbol) %>%
    group_by( gsymbol ) %>%
    dplyr::summarize( mean = mean( value ), cv = log( ( sd( value ) / mean( value ) ) ) ) %>%
    mutate( sig = ifelse( gsymbol %in% sig, "sig+", "sig-" ) ) %>%
    mutate( expressed = ifelse( mean > 6, "exprs+", "exprs-" ) ) %>%
    arrange( sig ) %>%
    mutate( fit_res = resid( loess( cv ~ mean ) ) ) %>%
    mutate( lm_genes = ifelse( fit_res > 0, "lm+", "lm-" ) ) %>%
    mutate( label = paste0( expressed, sig, lm_genes ) )
} )

signatures <- lapply( plotdat_sig_filt_l, function( df ) {
  df %>% filter( label == "exprs+sig+lm+" ) %>%
    mutate_all( funs( as.character ) ) %>%
    pull( gsymbol )
} )

plotdat <- bind_rows( plotdat_sig_filt_l, .id = "signature" )
plotdat$signature <- factor( plotdat$signature,
                            levels = c( "HRAS", "KRASG13D134", "KrasLA", "RAS_pathway", "MSigDB", "RAS_addiction" ) )

gg <- ggplot( plotdat, aes( x = mean, y = cv, color = label ) ) +
  geom_point( alpha = 0.3 ) +
  geom_smooth( method = "loess", aes( color = "line" ) ) +
  geom_point( data = subset( plotdat, label == "exprs+sig+lm+" ), size = 2,
             aes( x = mean, y = cv, fill = label ), colour = "black", pch = 21 ) +
  geom_point( data = subset( plotdat, label %in% c( "exprs-sig+lm-", "exprs-sig+lm+", "exprs+sig+lm-" ) ),
             size = 1 ) +
  scale_color_manual( values = c( 'exprs-sig-lm-' = my_grey, 'exprs-sig-lm+' = my_grey,
                                  'exprs-sig+lm-' = "black", 'exprs-sig+lm+' = "black",
                                  'exprs+sig-lm-' = my_grey, 'exprs+sig-lm+' = my_grey,
                                  'exprs+sig+lm-' = "black", 'exprs+sig+lm+' = my_grey,
                                 line = "black" ) ) +
  scale_fill_manual( values = c( 'exprs-sig-lm-' = my_grey, 'exprs-sig-lm+' = my_grey,
                                 'exprs-sig+lm-' = "black", 'exprs-sig+lm+' = "black",
                                 'exprs+sig-lm-' = my_grey, 'exprs+sig-lm+' = my_grey,
                                 'exprs+sig+lm-' = "black", 'exprs+sig+lm+' = my_green ) ) +
  plot_formatter( ) +
  theme( legend.position = "none" ) +
  labs( x = "mean log2 expression estimate", y = "log coefficient of variance" ) +
  facet_wrap( ~signature ) +
  theme( strip.background = element_blank( ),
         text = element_text( size = 10 ) )
print( gg )
```

Here we add the pre-computed RAS84 signature, derived below, to the
signature list so we can present it in the contingency table
figure.

```{r add_ras84}
ras84_feature_map <- read.xlsx( RAS84_FEATURE_MAP_FILE )
signatures$RAS84 <- ras84_feature_map$CCLE_feature_id
```

# Cluster cell lines using clean signatures {.tabset .tabset-fade}

Using the filtered signatures and cell-lines we clustered the cell-lines into three
groups and label them as high, medium and low based on the average
expression of the signature genes within each cluster. The clustering of the cell-lines for each signture are shown by the heatmaps below. KRAS mutation status is highlighted by the purple bars on the right hand side.

```{r cluster,results='asis',echo=FALSE}
signatures_df <- melt( signatures ) %>%
  unite( gene_sig, value, L1, remove = FALSE )

annodat <- fData( eset_ccle_lung_kras ) %>%
    rownames_to_column( )

samp_clust <- melt( lapply( signatures, function( sig ) {
  ## - extract expression values for signature
  dat <- exprs( eset_ccle_lung_kras ) %>%
      as.data.frame( ) %>%
      rownames_to_column( ) %>%
      left_join( annodat, by = "rowname" ) %>%
      filter( Description %in% sig ) %>%
      dplyr::select( -rowname, -Name, -Description ) %>%
      t( )
  ## - cluster and extract cluster membership
  clust <- hclust( dist( dat ), method = "ward.D2" )
  ct <- cutree( clust, 3 ) %>%
    as.data.frame( ) %>%
    select_( ct = "." ) %>%
    rownames_to_column( )
  ## - calculate mean cluster expression and relabel
  dat %>%
    as.data.frame( ) %>%
    rownames_to_column( ) %>%
    left_join( ct, by = "rowname" ) %>%
    tidyr::gather( name, value, -rowname, -ct ) %>%
    group_by( ct ) %>%
    dplyr::summarize( cluster_mean = mean( value ) ) %>%
    arrange( cluster_mean ) %>%
    mutate( rasact = c( "RASlow", "RASmed", "RAShigh" ) ) %>%
    right_join( ct, by = "ct" ) %>%
    dplyr::select( rasact, rowname )
} ) )

hmcols <- colorpanel( 1000, my_blue, "white", my_red )

for( sig_n in names( signatures ) ) {
    cat( "## ", sig_n, "\n" )
    hmdat <- exprs( eset_ccle_lung_kras ) %>%
        as.data.frame( ) %>%
        rownames_to_column( ) %>%
        left_join( annodat, by = "rowname" ) %>%
        filter( Description %in% signatures[[ sig_n ]] )
    hmmat <- hmdat %>% dplyr::select( -Name, -Description, -rowname )
    hmmat <- hmmat - rowMeans( hmmat )
    clusters <- samp_clust %>%
        filter( L1 == sig_n )
    hm_anno_df <- pData( eset_ccle_lung_kras ) %>%
        left_join( clusters, by = "rowname" ) %>%
        mutate( KRASmut = ifelse( onco_ras == "kras_mut", "+", "-" ) )
    hm_anno_row <- rowAnnotation( df = hm_anno_df[, "KRASmut", drop = FALSE ],
                                 col = list( KRASmut = c( "-" =  "white", "+" = my_purple ) ),
                                 annotation_legend_param = list( title_gp = gpar( fontsize = 14 ),
                                                                labels_gp = gpar( fontsize = 12 ) ),
                                 show_legend = FALSE,
                                 show_annotation_name = FALSE )
    hm3d <- Heatmap( t( hmmat ),
                    name = "Expression", 
                    clustering_method_rows = "ward.D2",
                    clustering_method_columns = "ward.D2",
                    show_column_names = FALSE,
                    show_row_names = FALSE,
                    col = colorRamp2(
                        seq( from = -3, to = 3, length.out = length( hmcols ) ),
                        hmcols ),
                    split = hm_anno_df$rasact,
                    column_title = sig_n,
                    column_title_gp = gpar( fontsize = 14 ),
                    heatmap_legend_param = list( title_gp = gpar( fontsize = 14 ),
                                                labels_gp = gpar( fontsize = 4 ) ),
                    show_heatmap_legend = FALSE,
                    cluster_row_slices = FALSE )
    pdf_file <- file.path( "figures", sub( "SIG", sig_n, SIGNATURE_HEATMAP_FILE ) )
    cairo_pdf( file = pdf_file, width = 2, height = 3 )
    print( hm3d + hm_anno_row )
    dev.off( )
    if( sig_n == "RAS84" ) {
        if( !file.exists( RAS84_HEATMAP_FILE ) ) {
            cairo_pdf( file = RAS84_HEATMAP_FILE, width = 3, height = 4 )
            print( hm3d + hm_anno_row )
            dev.off( )
        }
    }
    cat( " \n\n" )
}
```

# Signature assessment via KRAS mutation distribution

We assessed each signature by its ability to discriminate KRAS mutation
between the high and low groups. Here we use a chisq-test to assign
significance to the distribution of KRAS mutants across the high and
low groups. The contingency tables for each signture are shown in the 
heatmap below. We also determined the difference in mean expression between the high and low RAS activity groups and the ratio of classified samples (those classified as either high or low) compared to the total number of cell lines. These values are displayed for all signatures in the volcano plot below.

We observed that Loboda, Bild Nevins and Downard 146 out-performed the
other signatures in terms of KRAS mutation segregation. Downward 74
was discarded due to the low number of classified samples. We included
a number of related control signatures in the analysis to control for
a possible background effect. As observed the control signatures do
not show a significant association of KRAS mutation with the highly
transcribed cell-line group. Sign-Settleman is of particular interest
here since it is a RAS addiction signature and therefore RAS related.

```{r chisq}
run_chisq <- function( df ) {
    tab <- df[ , c( "kras_mut", "oncoras_neg" ) ]
    res <- chisq.test( as.table( as.matrix( tab ) ) )
    data.frame( lods =  odds <- log2( ( ( tab[ 1, 1 ]+0.1 ) / ( tab[ 3, 2 ] + 0.1 ) ) / ( ( tab[ 3, 1 ] + 0.1 ) / ( tab[ 1, 2 ] + 0.1 ) ) ),
                pval = res$p.value )
}

tab_dat <- samp_clust %>%
  left_join( pData( eset_ccle_lung_kras )[, c( "rowname", "onco_ras" ) ], by = "rowname" ) %>%
  filter( rasact != "RASmed" ) %>%
  group_by( L1, rasact, onco_ras ) %>%
  dplyr::summarize( n = n() ) %>%
  spread( onco_ras, n, fill = 0 )
    
chisq_res <- tab_dat %>%
    group_by( L1 ) %>%
    do( run_chisq( . ) )

hmcols <- colorpanel( 100, "white", "#E3001A" )

sigs <- c( "Loboda", "Bild.Nevins", "Downward_134",
          "KRAS_msigdb_HM", "SweetCordero", "Singh_Settleman", "RAS84" )

ord_df <- data.frame( signatures = sigs,
                      order = 1:7 )

hm_tab_dat <- tab_dat %>%
  filter( L1 %in% sigs ) %>%
  as.data.frame( ) %>%
  left_join( ord_df, by = c( "L1" = "signatures" ) ) %>%
  arrange( order ) %>%
  unite( signature, L1, rasact, remove = FALSE ) %>%
  dplyr::select( signature, rasact, kras_mut, oncoras_neg, L1 )
  
hmdat <- melt( hm_tab_dat ) %>%
  mutate( L1 = factor( signature, levels = hm_tab_dat$signature ) ) %>%
    mutate( value = ifelse( rasact == "RASlow", value*-1, value ) ) %>%
    mutate( variable = factor( variable, levels = c( "oncoras_neg", "kras_mut" ) ) )
hmdat$signature <- sig_name_map[ as.character( hmdat$signature ) ]

gg_hm <- ggplot( hmdat, aes( L1, variable ) ) +
  geom_tile( aes( fill = value ), colour = "white" ) +
  geom_text( aes( label = abs( value ) ) ) +
  scale_fill_gradient2( low = my_blue, mid = "white", high = my_red ) +
  labs(x = "", y = "") +
  scale_x_discrete( expand = c( 0, 0 ) ) +
  scale_y_discrete(expand = c( 0, 0 ) ) +
  theme( panel.spacing = unit( c( 0, 0, 0, 0 ), "cm" ),
        axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        axis.title.y=element_blank(),
        axis.ticks.y=element_blank(),
        axis.ticks.length = unit(0, "pt"),
        panel.border = element_blank( ),
        legend.position = "none" )

ras_act_df <- samp_clust %>%
  unite( sample_sig_map, L1, rowname, remove = FALSE )

rasact_group_exprs <- exprs( eset_ccle_lung_kras ) %>%
  as.data.frame( ) %>%
  rownames_to_column( ) %>%
  left_join( annodat, by = "rowname" ) %>%
  filter( Description %in% signatures_df$value ) %>%
  tidyr::gather( name, value, -rowname, -Name, -Description ) %>%
  left_join( signatures_df, by = c( "Description" = "value" ) ) %>%
  unite( sample_sig_map, L1, name, remove = FALSE ) %>%
  full_join( ras_act_df, by = "sample_sig_map" ) %>%
  group_by( L1.y, rasact ) %>%
  dplyr::summarize( mean = mean( value ) )

rasact_group_lfc <- rasact_group_exprs %>%
  spread( rasact, mean ) %>%
  mutate( l2fc = RAShigh - RASlow )

plotdat <- exprs( eset_ccle_lung_kras ) %>%
  as.data.frame( ) %>%
  rownames_to_column( ) %>%
  left_join( annodat, by = "rowname" ) %>%
  tidyr::gather( name, value, -rowname, -Name, -Description ) %>%
  filter( Description %in% signatures_df$value ) %>%
  left_join( samp_clust, by = c( "name" = "rowname" ) ) %>%
  filter( L1 %in% sigs ) %>%
  filter( rasact %in% c( "RASlow", "RAShigh" ) ) %>%
  unite( gene_sig, Description, L1, remove = FALSE ) %>%
  filter( gene_sig %in% signatures_df$gene_sig ) %>%
  mutate( signatures = factor( sig_name_map[ as.character( L1 ) ], levels = sig_name_map[ sigs ] ) )

gg_box <- ggplot( plotdat, aes( x = signatures, y = value, fill = rasact, color = rasact ) ) +
    geom_boxplot( alpha = 0.4, width = 0.4, position = position_dodge( 0.6 ) ) +
    scale_fill_manual( values = c( my_red, my_blue ) ) +
    scale_color_manual( values = c( my_red, my_blue ) ) +
    labs( x = "log2 expression" ) +
    plot_formatter( ) +
    theme( legend.position = "none",
          axis.text.x = element_text( angle = 90, hjust = 1, vjust = 0.5 ) ) +
    geom_vline( xintercept = (1:6) + 0.5, linetype = "longdash", color = my_grey ) +
    coord_cartesian( xlim = c( 0.5, 7.5 ), ylim = c( 2, 15 ), expand = FALSE ) +
    labs( y = "log2 exprs", x = "RAS activity groups" ) +
    theme( axis.text.x = element_text( angle = 35, hjust = 1, vjust = 1 ) )

png( file = file.path( adat$plot.path, "gg_box.png" ) )
print( gg_box )
dev.off()

## - merge contingency heatmap and boxplots
cairo_pdf( file = RAS_HIGHLOW_CONTINGENCY_FILE, height = 3, width = 6.5 )
ggs <- list( gg_hm = gg_hm, gg_box = gg_box )
grobL <- lapply( ggs, ggplotGrob )
widths <- lapply( grobL, function( gb ) gb$widths[ 2:5 ] )
widthCol1 <- do.call( grid::unit.pmax, widths[ names( ggs ) ] )
grobL$gg_hm$widths[ 2:5 ] <- as.list( widthCol1 )
grobL$gg_box$widths[ 2:5 ] <- as.list( widthCol1 )
gt <- gtable( unit( 1, "npc" ),
             unit( c( 0.2, 0.8 ), "npc" ),
             respect = TRUE )
gt <- gtable_add_grob( gt, grobL$gg_box, 2, 1 )
gt <- gtable_add_grob( gt, grobL$gg_hm, 1, 1 )
grid.draw( gt )
dev.off( )

print( grid.draw( gt ) )
```
```{r remove_ras84}
signatures <- signatures[ -7 ]
```

## Volcano plot showing signature performance

Here we create a volcano plot highlighting each signature's ability to segregate KRAS mutations across the RAS high and low groups, the differences in the log2 mean expression values between the two groups and the ratio of the number of samples classified as high or low.

```{r ras_volcano}
tab_dat <- samp_clust %>%
  left_join( pData( eset_ccle_lung_kras )[, c( "rowname", "onco_ras" ) ], by = "rowname" ) %>%
    dplyr::rename( signatures = L1 ) %>%
  mutate( signatures = factor( signatures ) ) %>%
  mutate( onco_ras_num = as.numeric( factor( onco_ras, levels = c( "oncoras_neg", "kras_mut" ) ) ) - 1 )

tab_dat_l <- split( tab_dat, tab_dat$signatures )

chisq <- map( tab_dat_l, function( df ) {
    chisq.test( table( df[ , c( "rasact", "onco_ras" ) ] ) )$p.value
} )

glm_fit <- map( tab_dat_l, function( df ) {
    glm( onco_ras_num ~ rasact, data = df, family = "binomial" )
} )

plotdat <- map( glm_fit, logLik )%>%
    melt( ) %>%
    dplyr::rename( signatures = L1 )

loglik_gg <- ggplot( plotdat, aes( x = 1, y = value, label = signatures ) ) +
    geom_vline( xintercept = 1 ) +
    geom_point( size = 5 ) +
    geom_text_repel( nudge_x = 0.1 ) +
    plot_formatter( ) +
    theme( legend.position = "none",
          text = element_text( size = 8 ) )

cairo_pdf( file = KRAS_MUTATION_LOGLIK_FILE, width = 3, height = 3 )
print( loglik_gg )
dev.off( )
```

# Combined 'RAS pathway', HRAS and KRASG13D134 signature heatmap

We constructed our own meta-signature from the three best performing
signatures, Loboda, Bild-nevins, Downward146. Here we pooled the
signature genes, reran the clustering and then enriched for genes
driving the clustering by selecting genes that are significantly
differential between the high and low groups. The pooled signature clustering is shown in the first heatmap below. RAS activity dependent gene selection is highlighted in the volcano plots below and the clustering of the contructed RAS84 signature is shown in the second heatmap.

```{r ras84}
hmcols <- colorpanel( 1000, my_blue, "white", my_red )

RAS84_parent_signatures <- c( "RAS_pathway", "HRAS", "KRASG13D134" )
RAS84_raw <- unique( unlist( signatures[ RAS84_parent_signatures ] ) )

hmdat <- exprs( eset_ccle_lung_kras ) %>%
    as.data.frame( ) %>%
    rownames_to_column( ) %>%
    left_join( annodat, by = "rowname" ) %>%
    filter( Description %in% RAS84_raw )

parents <- melt( signatures[ RAS84_parent_signatures ], value.name = "Description" ) %>%
    filter( Description %in% RAS84_raw ) %>%
    mutate( flag = TRUE ) %>%
    spread( L1, flag, fill = FALSE )

hmdat <- hmdat %>%
    left_join( parents, by = "Description" )

hmmat <- hmdat %>%
    dplyr::select( -Name, -Description, -rowname ) %>%
    dplyr::select_( .dots = paste0( "-", RAS84_parent_signatures ) )

hmmat <- hmmat - rowMeans( hmmat )

hm_anno_df <- pData( eset_ccle_lung_kras ) %>%
    mutate( KRASmut = ifelse( onco_ras == "kras_mut", "+", "-" ) )
hm_anno_row <- rowAnnotation( df = hm_anno_df[, "KRASmut", drop = FALSE ],
                             col = list( KRASmut = c( "-" =  "white", "+" = my_purple ) ),
                             annotation_width = 1,
                             annotation_legend_param = list( title_gp = gpar( fontsize = 14 ),
                                                            labels_gp = gpar( fontsize = 12 ) ),
                             show_legend = FALSE )

col_anno_dat <- dplyr::select_( hmdat, .dots = RAS84_parent_signatures )
col_anno_colours <- lapply( colnames( col_anno_dat ), function( x ) c( 'TRUE' = "lightgrey", 'FALSE' = "white" ) )
names( col_anno_colours ) <- colnames( col_anno_dat )

hm_anno_col <- HeatmapAnnotation( df = col_anno_dat,
                                 show_annotation_name = TRUE,
                                 simple_anno_size = unit( 1, "mm" ),
                                 col = col_anno_colours,
                                 annotation_width = 0.01,
                                 na_col = "white",
                                 gap = unit( 1, "mm" ),
                                 show_legend = FALSE,
                                 annotation_name_gp = gpar( fontsize = 6 ),
                                 annotation_name_side = "left" )

hm3d <- Heatmap( t( hmmat ),
                name = "Expression", 
                clustering_method_rows = "ward.D2",
                clustering_method_columns = "ward.D2",
                show_column_names = FALSE,
                show_row_names = FALSE,
                col = colorRamp2(
                    seq( from = -3, to = 3, length.out = length( hmcols ) ),
                    hmcols ),
                split = 3,
                column_title = "RAS84_raw",
                column_title_gp = gpar( fontsize = 14 ),
                heatmap_legend_param = list( title_gp = gpar( fontsize = 14 ),
                                             labels_gp = gpar( fontsize = 4 ) ),
                show_heatmap_legend = FALSE,
                bottom_annotation = hm_anno_col )
print( hm3d + hm_anno_row )
if( !file.exists( RAS84_RAW_HEATMAP) ) ) {
    cairo_pdf( file = RAS84_RAW_HEATMAP, width = 2, height = 3 )
    print( hm3d + hm_anno_row )
    dev.off( )
}
```

# RAS84 signature gene selection

Here we identify cluster driver genes via statistical analysis across the clusters. Cluster dependent gene expression is identified using limma and a 0.05 fdr threshold. We construct RAS84 from the genes significant in the RAS high group in signatures RAS pathway, HRAS and KRASG13D134.

```{r limma}
sigs_drivers<- sapply( RAS84_parent_signatures, function( sig_n ) {
  clusters <- samp_clust %>%
    filter( L1 == sig_n )
  pdat <- pData( eset_ccle_lung_kras ) %>%
    left_join( clusters, by = "rowname" ) %>%
    mutate( rasact = factor( rasact, levels = c( "RASlow", "RASmed", "RAShigh" ) ) )
  model <- model.matrix( ~rasact, pdat )
  eset_test <- eset_ccle_lung_kras[ fData( eset_ccle_lung_kras )$Description %in% signatures[[ sig_n ]], ]
  fit <- eBayes( lmFit( eset_test, model ) )
  ind <- decideTests( fit, adjust.method = "fdr", p.value = 0.05 )[, "rasactRAShigh" ] == 1
  data.frame(
    decide = ind,
    pval = fit$p.value[ , "rasactRAShigh" ],
    coef = fit$coef[ , "rasactRAShigh" ],
    genes = as.character( fData( eset_test )$Description ) )
}, simplify = FALSE )

signatures$RAS84 <- unique( as.character( unlist( lapply( sigs_drivers, function( df ) df$genes[ df$decide ] ) ) ) )

## - add RAS84 index to pData
fdat <- fData( eset_ccle_lung_kras ) %>%
  rownames_to_column( )

RAS84_index_df <- exprs( eset_ccle_lung_kras ) %>%
  as.data.frame( ) %>%
  rownames_to_column( ) %>%
  left_join( fdat, by = "rowname" ) %>%
  filter( Description %in% signatures$RAS84 ) %>%
  tidyr::gather( sample.id, value, -rowname, -Name, -Description ) %>%
  group_by( sample.id ) %>%
  dplyr::summarize( RAS84_index = mean( value ) )

eset_ccle_lung_kras$RAS84_index <- pData( eset_ccle_lung_kras ) %>%
  left_join( RAS84_index_df, by = "sample.id" ) %>%
  pull( RAS84_index )
```

# Volcano plots for signature gene expression across clusters

Here we show the volcano plots from the cluster limma analysis above.

```{r sig_volcanos}
plotdat <- melt( sigs_drivers ) %>%
    spread( variable, value ) %>%
    mutate( log10pval = -log10( pval ) ) %>%
    mutate( color = paste( L1, decide, sep = "-" ) ) %>%
    mutate( signatures = factor( L1, levels = RAS84_parent_signatures ) )
plotdat$signatures <- sig_name_map[ as.character( plotdat$signatures ) ]

gg <- ggplot( plotdat, aes( x = coef, y = log10pval, fill = decide ) ) +
    geom_point( colour = "black", pch = 21, size = 2 ) +
    geom_vline( xintercept = 0, linetype = "dashed" ) +
    facet_wrap( ~signatures, ncol = 3 ) +
    scale_fill_manual( values = c( 'FALSE' = my_grey, 'TRUE' = my_green ) ) +
    plot_formatter( ) +
    theme( legend.position = "none",
          strip.background = element_blank( ),
          text = element_text( size = 10 ) ) +
    labs( x = "log2 fold change", y = "-log10 p-value" )
print( gg )

if( !file.exists( SIGNATURE_CLUSTER_VOLCANOS_FILE ) {
    cairo_pdf( file = SIGNATURE_CLUSTER_VOLCANOS_FILE, width = 3.5, height = 1.75)
    print( gg )
    dev.off( )
}
```

# RAS84 signature heatmap

RAS84 signature heatmap showing cell-line KRAS mutation status across
the RAS activity groups. Parent signature gene membership is also
shown in the column annotation at the bottom of the heatmap.

```{r ras84_heatmap}
sig_f <- fData( eset_ccle_lung_kras )$Description %in% signatures$RAS84

dat <- exprs( eset_ccle_lung_kras ) %>%
    as.data.frame( ) %>%
    rownames_to_column( ) %>%
    left_join( annodat, by = "rowname" ) %>%
    filter( Description %in% signatures$RAS84 ) %>%
    dplyr::select( -rowname, -Name, -Description )
clust <- hclust( dist( t( dat ) ), method = "ward.D2" )
clusters <- cutree( clust, 3 )

pData( eset_ccle_lung_kras )$RAS84 <- c( "RASlow", "RASmed", "RAShigh" )[ clusters ]
pData( eset_ccle_lung_kras )$RAS84 <- factor( pData( eset_ccle_lung_kras )$RAS84,
                                             levels = c( "RAShigh", "RASlow", "RASmed" ) )
saveRDS( eset_ccle_lung_kras, "eset_ccle_lung_kras.rds" )

active_sig <- "RAS84"
hmdat <- exprs( eset_ccle_lung_kras ) %>%
    as.data.frame( ) %>%
    rownames_to_column( ) %>%
    left_join( annodat, by = "rowname" ) %>%
    filter( Description %in% signatures$RAS84 )

parents <- melt( signatures[ RAS84_parent_signatures ], value.name = "Description" ) %>%
    filter( Description %in% signatures[[ active_sig ]] ) %>%
    mutate( flag = TRUE ) %>%
    spread( L1, flag, fill = FALSE )

hmdat <- hmdat %>%
    left_join( parents, by = "Description" )

hmmat <- hmdat %>%
    column_to_rownames( "Description" ) %>%
    dplyr::select( -Name, -rowname ) %>%
    dplyr::select_( .dots = paste0( "-", RAS84_parent_signatures ) )

hmmat <- hmmat - rowMeans( hmmat )

clusters <- samp_clust %>%
    filter( L1 == active_sig )
hm_anno_df <- pData( eset_ccle_lung_kras ) %>%
    left_join( clusters, by = "rowname" ) %>%
    mutate( KRASmut = ifelse( onco_ras == "kras_mut", "+", "-" ) )
hm_anno_row <- rowAnnotation( df = hm_anno_df[, "KRASmut", drop = FALSE ],
                             col = list( KRASmut = c( "-" =  "white", "+" = my_purple ) ),
                             annotation_width = 1,
                             annotation_legend_param = list( title_gp = gpar( fontsize = 14 ),
                                                            labels_gp = gpar( fontsize = 12 ) ),
                             show_legend = FALSE,
                             show_annotation_name = FALSE,
                             simple_anno_size = unit( 3, "mm" ) )

col_anno_dat <- dplyr::select_( hmdat, .dots = RAS84_parent_signatures )
colnames( col_anno_dat ) <- sig_name_map[ colnames( col_anno_dat ) ]
col_anno_colours <- lapply( colnames( col_anno_dat ), function( x ) c( 'TRUE' = "lightgrey", 'FALSE' = "white" ) )
names( col_anno_colours ) <- colnames( col_anno_dat )

hm_anno_col <- HeatmapAnnotation( df = col_anno_dat,
                                 show_annotation_name = TRUE,
                                 simple_anno_size = unit( 1, "mm" ),
                                 col = col_anno_colours,
                                 annotation_width = 0.01,
                                 na_col = "white",
                                 gap = unit( 1, "mm" ),
                                 show_legend = FALSE,
                                 annotation_name_gp = gpar( fontsize = 6 ),
                                 annotation_name_side = "left" )

hm3d <- Heatmap( t( hmmat ),
                name = "Expression", 
                clustering_method_rows = "ward.D2",
                clustering_method_columns = "ward.D2",
                show_column_names = TRUE,
                show_row_names = FALSE,
                col = colorRamp2(
                    seq( from = -3, to = 3, length.out = length( hmcols ) ),
                    hmcols ),
                row_split = hm_anno_df$RAS84,
                column_title = active_sig,
                column_title_gp = gpar( fontsize = 14 ),
                column_names_gp = gpar( fontsize = 4 ),
                heatmap_legend_param = list( title_gp = gpar( fontsize = 14 ),
                                            labels_gp = gpar( fontsize = 4 ) ),
                show_heatmap_legend = TRUE,
                bottom_annotation = hm_anno_col,
                cluster_row_slices = FALSE,
                row_names_max_width = unit( 10, "mm" ) )
print( hm3d + hm_anno_row )

if( !file.exists( RAS84_SIGNATURE_HEATMAP_FILE ) ) {
    cairo_pdf( file = RAS84_SIGNATURE_HEATMAP_FILE, width = 6, height = 4 )
    print( hm3d + hm_anno_row )
    dev.off( )
}
```

# RAS84 signature volcano plot

The performance of all signatures are highlighted in this volcano plot.
The significance of KRAS mutation segregation across RAS activity high
and low groups is shown on the y-axis. The log2 fold change in average
expression between the high and low groups is shown on the x-axis. The
point size reflects the ratio of high+low classified samples.

```{r volcano}
color_map <- data.frame( sig_n = names( signatures ),
                         group = c( "selected", "selected",
                                   "discarded", "selected", "control", "discarded",
                                   "merged" ) )

rasact_group_counts <- ras_act_df %>%
  group_by( L1, rasact ) %>%
  dplyr::summarize( n = n( ) ) %>%
  spread( rasact, n ) %>%
  mutate( ratio = ( RAShigh + RASlow ) / ( RAShigh + RASmed + RASlow ) * 100 )

plotdat_sig_volc <- rasact_group_lfc %>%
    filter( L1.y %in% color_map$sig_n ) %>%
    left_join( chisq_res, by = c( "L1.y" = "L1" ) ) %>%
    left_join( rasact_group_counts, by = c( "L1.y" = "L1" ) ) %>%
    left_join( color_map, by = c( "L1.y" = "sig_n" ) ) %>%
    mutate( log10pval = -log10( pval ) )  %>%
    filter( !L1.y %in% c( "RAS47" ) ) %>%
    mutate( group = factor( group, levels = c( "merged", "selected", "discarded", "control" ) ) ) %>%
    dplyr::rename( signatures = L1.y )

plotdat_sig_volc$signatures <- sig_name_map[ as.character( plotdat_sig_volc$signatures ) ]
gg1d_all_sigs <- ggplot( plotdat_sig_volc, aes( x = l2fc, y = log10pval, label = signatures, size = ratio, color = group ) ) +
  geom_point() +
  geom_label_repel( size = 3,  nudge_x = 0.1, nudge_y = 0.3,
                   show.legend = FALSE, inherit.aes = FALSE,
                   data = plotdat_sig_volc, aes( x = l2fc, y = log10pval, label = signatures, color = group ) ) +
    geom_hline( yintercept = -log10( 0.05 ), lty = "dotted" ) +
    labs( x = "log2FoldChange", y = "-log10 p-value", size = "% highlow" ) +
    plot_formatter( ) +
    scale_color_manual( values = c( my_purple, my_blue, my_green, my_grey ) )
    

gg1d_all_sigs <- gg1d_all_sigs + theme( legend.position = "right" )
print( gg1d_all_sigs )

if( !file.exists( SIGNATURE_VOLCANO_YES_RAS84_FILE ) ) {
    cairo_pdf( file = SIGNATURE_VOLCANO_YES_RAS84_FILE, width = 5, height = 4 )
    gg1d_all_sigs
    dev.off( )
}

new_pvalfig_gg <- ggplot( plotdat_sig_volc,
                         aes( x = 1, y = -log10( pval ),
                             color = group, label = signatures ) ) +
    geom_vline( xintercept = 1 ) +
    geom_point( size = 5 ) +
    geom_text_repel( nudge_x = 0.1 ) +
    plot_formatter( ) +
    scale_color_manual( values = c( my_purple, my_blue, my_green, my_grey ) ) +
    theme( legend.position = "none",
           text = element_text( size = 8 ) ) 

cairo_pdf( file = KRAS_MUTATION_PVALUE_FILE, width = 3, height = 3 )
print( new_pvalfig_gg )
dev.off( )
```

# Control signatures

Here we take a look at how control signatures perform using our RAS activity validation strategy. Control signatures are onco-genic pathway signatures that are related to RAS signally but are not expected to drive exactly the same transcriptional response.

```{r controls}
control_sig_files <- dir( path = CONTROL_SIGNATURE_HOME, pattern = "\\.txt" )
control_sigs <- sapply( control_sig_files, function( f ) {
  read.delim( file = file.path( control_sig_home, f ),
             stringsAsFactors = FALSE, header = FALSE )[, 1 ]
}, simplify = FALSE )
names( control_sigs ) <- sub( "\\.txt", "", names( control_sigs ) )
control_sigs_df <- melt( control_sigs ) %>%
  unite( gene_sig, value, L1, remove = FALSE )

annodat <- fData( eset_ccle_lung_kras ) %>%
    rownames_to_column( )

samp_clust <- melt( lapply( control_sigs, function( sig ) {
  ## - extract expression values for signature
  dat <- exprs( eset_ccle_lung_kras ) %>%
      as.data.frame( ) %>%
      rownames_to_column( ) %>%
      left_join( annodat, by = "rowname" ) %>%
      filter( Description %in% sig ) %>%
      dplyr::select( -rowname, -Name, -Description ) %>%
      t( )
  ## - cluster and extract cluster membership
  clust <- hclust( dist( dat ), method = "ward.D2" )
  ct <- cutree( clust, 3 ) %>%
    as.data.frame( ) %>%
    select_( ct = "." ) %>%
    rownames_to_column( )
  ## - calculate mean cluster expression and relabel
  dat %>%
    as.data.frame( ) %>%
    rownames_to_column( ) %>%
    left_join( ct, by = "rowname" ) %>%
    tidyr::gather( name, value, -rowname, -ct ) %>%
    group_by( ct ) %>%
    dplyr::summarize( cluster_mean = mean( value ) ) %>%
    arrange( cluster_mean ) %>%
    mutate( rasact = c( "RASlow", "RASmed", "RAShigh" ) ) %>%
    right_join( ct, by = "ct" ) %>%
    dplyr::select( rasact, rowname )
} ) )

run_chisq <- function( df ) {
    tab <- df[ , c( "kras_mut", "oncoras_neg" ) ]
    res <- chisq.test( as.table( as.matrix( tab ) ) )
    data.frame( lods =  odds <- log2( ( ( tab[ 1, 1 ]+0.1 ) / ( tab[ 3, 2 ] + 0.1 ) ) / ( ( tab[ 3, 1 ] + 0.1 ) / ( tab[ 1, 2 ] + 0.1 ) ) ),
                pval = res$p.value )
}

tab_dat <- samp_clust %>%
  left_join( pData( eset_ccle_lung_kras )[, c( "rowname", "onco_ras" ) ], by = "rowname" ) %>%
  filter( rasact != "RASmed" ) %>%
  group_by( L1, rasact, onco_ras ) %>%
  dplyr::summarize( n = n() ) %>%
  spread( onco_ras, n, fill = 0 )
    
chisq_res <- tab_dat %>%
    #dplyr::select( -value ) %>%
    group_by( L1 ) %>%
    do( run_chisq( . ) )

ras_act_df <- samp_clust %>%
  unite( sample_sig_map, L1, rowname, remove = FALSE )

rasact_group_exprs <- exprs( eset_ccle_lung_kras ) %>%
  as.data.frame( ) %>%
  rownames_to_column( ) %>%
  left_join( annodat, by = "rowname" ) %>%
  filter( Description %in% control_sigs_df$value ) %>%
  tidyr::gather( name, value, -rowname, -Name, -Description ) %>%
  left_join( control_sigs_df, by = c( "Description" = "value" ) ) %>%
  unite( sample_sig_map, L1, name, remove = FALSE ) %>%
  full_join( ras_act_df, by = "sample_sig_map" ) %>%
  group_by( L1.y, rasact ) %>%
  dplyr::summarize( mean = mean( value ) )

rasact_group_lfc <- rasact_group_exprs %>%
  spread( rasact, mean ) %>%
  mutate( l2fc = RAShigh - RASlow )

rasact_group_counts <- ras_act_df %>%
  group_by( L1, rasact ) %>%
  dplyr::summarize( n = n( ) ) %>%
  spread( rasact, n ) %>%
  mutate( ratio = ( RAShigh + RASlow ) / ( RAShigh + RASmed + RASlow ) * 100 )

color_map_controls <- data.frame( sig_n = names( control_sigs ),
                                  group = "control" )

plotdat_controls <- rasact_group_lfc %>%
    left_join( chisq_res, by = c( "L1.y" = "L1" ) ) %>%
    left_join( rasact_group_counts, by = c( "L1.y" = "L1" ) ) %>%
    mutate( group = "control" ) %>%
    mutate( log10pval = -log10( pval ) )

plotdat_controls <- data.frame( signatures = plotdat_controls$L1.y,
                                plotdat_controls )

plotdat <- bind_rows( plotdat_sig_volc, plotdat_controls )

gg1d_all_sigs <- ggplot( plotdat, aes( x = l2fc, y = log10pval,
                                       label = signatures, color = group ) ) +
  geom_point( size = 3) +
  geom_text_repel( size = 3,  nudge_x = 0.1, nudge_y = 0.3,
                   show.legend = FALSE, inherit.aes = FALSE,
                   data = plotdat, aes( x = l2fc, y = log10pval, label = signatures, color = group ) ) +
  geom_hline( yintercept = -log10( 0.05 ), lty = "dotted" ) +
  labs( x = "log2FoldChange", y = "-log10 p-value", size = "% highlow" ) +
  plot_formatter( ) +
  theme( text = element_text( size = 10 ) ) +
  scale_color_manual( values = c( my_grey, my_green, my_purple, my_blue) )

gg1d_all_sigs <- gg1d_all_sigs + theme( legend.position = "right" )
print( gg1d_all_sigs )
if( !file.exists( CONTROL_SIGNATURES_VOLCANO_FILE ) ) {
    cairo_pdf( file = CONTROL_SIGNATURES_VOLCANO_FILE

            , width = 5, height = 4 )
    gg1d_all_sigs
    dev.off( )
}
```
