---
title: "RAS84 CCLE Lung Cancer Cell Line Drug Sensitivity Analysis"
author: "philip.east@crick.ac.uk"
date: "25/03/2019"
output:
  html_document:
    df_print: tibble
    toc: true
    toc_depth: 5
    toc_float: true
    code_folding: hide
    number_sections: true
---

```{r setup,include=FALSE}
knitr::opts_chunk$set(echo = TRUE,message=FALSE,warning=FALSE)
```

```{r r_init,message=FALSE}
library( DESeq2 )
library( reshape2 )
library( ggplot2 )
library( rtracklayer )
library( RColorBrewer )
library( ComplexHeatmap )
library( openxlsx )
library( dplyr )
library( tibble )
library( DT )
library( ggrepel )
library( Seurat )
library( GenomicFeatures )
library( ggrepel )
library( tidyverse )
library( circlize )
library( GGally )
library( clusterProfiler )
library( org.Hs.eg.db )
library( bamsignals )
library( KernSmooth )
library( Seurat )
library( biomaRt )
library( gplots )
library( broom )
library( dendextend )
library( UpSetR )

options( max.print=300, width=100 )

source( "/camp/stp/babs/working/eastp/code/R/R.packages.east01.functions.R")
source( "../../scripts/lib.Ras_HighLow.335.R" )
source( "../../scripts/lib.CCLE.335.R" )

my_grey <- "#707173"
my_red <- "#e3001a"
my_orange <- "#f6ad6e"
my_green <- "#7ab51d"
my_lightgreen <- "#adcf82"
my_purple <- "#bb90bd"
my_blue <- "#4066aa"

plot_formatter <- function() {
    theme_bw( ) +
        theme( panel.grid.major = element_blank( ),
               panel.grid.minor = element_blank( ),
               panel.border = element_blank( ),
               panel.background = element_blank( ),
               axis.line = element_line(color = "black"),
               axis.line.x = element_line(color="black", size = 0.1 ),
               axis.line.y = element_line(color="black", size = 0.1 ),
               text = element_text( size = 10 ) )
}

adat <- project.init( "drug_screen", root = "results" )
```

# Introduction

To determine if any RAS84 lung cancer cell line group was sensitive/resistant to
any cancer drug we screened the groups against cancer cell
line drug sensitivity data from the GDSC and the CTRP. This analysis
is presented below. We first clustered the lung cell line RNA-Seq data
into two and five groups. We mapped mutation data to these groups to
see if we could recapitulate what we saw with patient samples from the
TCGA LUAD cohort. We compared IC50/AUC sensitivity values across the
groups to determine if any one group was more or less
sensitive to any of the compounds. We also compared KRAS mutant verse
wild type cell lines using GDSC data alone and the intersection
between the GDSC and CCLE cell lines. We also compared RAS pathway altered verses RAS pathway
wildtype lines that were present in both datasets. 

# Drug response metrics explained

## Dose-response curves

Drug response metrics are derived from a dose-response curve. The
dose–response relationship, or exposure–response relationship,
describes the magnitude of the response of an organism, as a function
of exposure (or doses) to a stimulus or stressor (usually a
chemical/drug) after a certain exposure time. 

In the context of cancer drugs, where the aim is to kill the cancer
cells, the response measure is cell viability. The dose-response curve
then becomes cell viability as a function of drug concentration.

[Dose-response](https://en.wikipedia.org/wiki/Dose%E2%80%93response_relationship)

## Metrics

### IC50

The half maximal inhibitory concentration (IC50)

The drug concentration to achieve 50% inhibition. Strictly speaking
this metric refers to the concentration of an antagonist required to
achieve 50% inhibition of the agonist. In the
case of the cell viability as a response variable it is the
concentration to achieve 50% cell viability.

![dose response curve](docs/Example_IC50_curve_demonstrating_visually_how_IC50_is_derived.png)

### EC50

Half maximal effective concentration (EC50)

The drug concentration to achieve half the maximum response. In cancer
drug cell viability studies this value would be calculated where the
response variable would be cell death.

### AUC

The area under the does response curve.

The AUC represents the total response over time and is used in
pharmacokinetics in patients. 

# Cancer drug sensitivity datasets

There are two large cancer drug sensitivity datasets available, the
GDSC and the CTRP, each detailed below.

## Genomics of Drug Sensitivity in Cancer (GDSC)

* A collaboration between The Sanger and Massachusettes General
Hospital Cancer Center.

* 518 compounds targeting 24 pathways across 988 cell lines.

* https://www.cancerrxgene.org

* GDSC1 987 Cell lines   367 Compounds
* GDSC2 809 Cell lines   198 Compounds

The GDSC data contains LN_IC50 and AUC drug sensitivity metrics. Data
summary statistics are shown below. Some drugs have been tested twice
on the same cell line in the GDSC data. In these cases, the lowest
LN_IC50 value was selected.

GDSC dose-response analysis was carried out using
[gdscIC50](https://github.com/CancerRxGene/gdscIC50) R package.

```{r gdsc_drug_data}
gdsc_se_l <- readRDS( file.path( "data", "gdsc_se_l.rds" ) )
```

## The Cancer Therapeutics Response Portal (CTRP)

* https://portals.broadinstitute.org/ctrp.v2.1/

* CTRP v1 242 CCLs 185 compounds
* CTRP v2 860 CCLs 481 compounds


### CTRPv1

CTRPv1 comes with AUC drug sensitivity values. We integrated cell line
and drug information from additional data files. The version 1
data is complied from assays of 354 drug compounds across 242 cell lines.

```{r ctrpv1_drug_data}
CTRPv1_se <- readRDS( file.path( "data", "CTRPv1_se.rds" ) )
```

### CTRPv2

CTRPv2 comes with AUC and EC50 drug sensitivity values. We integrated cell line
and drug information from additional data files. The version 2
data is complied from assays of 545 compounds across 887 cell lines.

```{r ctrpv2_drug_data}
CTRPv2_se <- readRDS( file.path( "data", "CTRPv2_se.rds" ) )
```

## Drug sensitivity dataset metrics

We looked at summary statistics associated with each of the
drug sensitivity datasets to assess group numbers and sensitivity
value distributions.

### Drug and cell line counts

```{r integrate_drug_resp}
drug_se_l <- c( gdsc_se_l, list( CTRPv1 = CTRPv1_se, CTRPv2 = CTRPv2_se ) )
bar_gg <- map( drug_se_l, dim ) %>%
    map( as.data.frame ) %>%
    map( rename, n = 1 ) %>%
    map( mutate, label = c( "drugs", "cell_lines" ) ) %>%
    bind_rows( .id = "dataset" ) %>%
    ggplot( aes( x = dataset, y = n, fill = label ) ) +
    geom_bar( stat = "identity", position = "dodge" ) +
    theme( axis.text.x = element_text( angle = 90, vjust = 0.5, hjust=1 ) )
print( bar_gg )
```

### Sensitivity metric distributions {.tabset .tabset-fade}

We looked at IC50 and AUC distributions across the datasets.

```{r drug_resp_dist}
drug_resp_gg_l <- map2( drug_se_l, names( drug_se_l ), function( se, dataset_n ) {
    map2( as.list( assays( se ) ), names( assays( se ) ), function( mat, data_n ) {
        mat %>% as.data.frame( ) %>%
            gather( ccl_name, value ) %>%
            filter( !is.na( value ) ) %>%
            ggplot( aes( x = value ) ) +
            geom_density( alpha = 0.3 ) +
            labs( title = data_n )
    } )
} )
```
```{r render_drug_resp_dist,results='asis',echo=FALSE}
for( i in names( drug_resp_gg_l ) ) {
    l <- drug_resp_gg_l[[ i ]]
    cat( "#### ", i, " {.tabset .tabset-fade}\n" )
    for( j in names( l )[ -1 ] ) {
        cat( "##### ", j, " \n" )
        print( l[[ j ]] )
        cat( " \n\n" )
    }
    cat( " \n\n" )
}
```

# RAS Signatures

We read in RAS signature gene lists.

```{r ras84_sig}
SIG_HOME <- file.path( "data", "signatures" )
krasSigNames <- sub( "\\.csv", "", dir( SIG_HOME ) )
signatures <- lapply( krasSigNames, function( sigFile ) {
    read.table( file = file.path( SIG_HOME, paste( sigFile, "csv", sep = "." ) ),
               sep = "\t", stringsAsFactors = FALSE )[, 1 ]
} )
names( signatures ) <- make.names( krasSigNames )

ras84_map <- read.xlsx( file.path( "data", "RAS84_feature_map.xlsx" ) )
```

# The CCLE data

RNA-Seq and mutation call data were obtained from the CCLE.

https://portals.broadinstitute.org/ccle/data


## RNA-Seq 

RNA-Seq data were obtained from the CCLE. Additional annotations were
obtained from Ensembl. We created a summarizedExperiment object from
these data.

## Mutation data

Mutation calls were obtained from CCLE. These were filtered for
RAS pathway genes as defined in [Oncogenic Signaling Pathways in The
Cancer Genome
Atlas](https://www.sciencedirect.com/science/article/pii/S0092867418303593?via%3Dihub)
and cancer driver mutations reported in [Comprehensive Characterization of Cancer Driver Genes and Mutations](https://www.sciencedirect.com/science/article/pii/S009286741830237X?via%3Dihub).
The data reports mutations at the transcript level so gene level
annotation was obtained from Ensembl and the data collapsed to the gene level.

```{r setup_se}
rasPathwayDefFile <- file.path( "..", "..", "data", "oncogenicRASPath_TCGA", "rasPathwayDefinition.txt" )
rasPathwayTab <- read.delim( file = rasPathwayDefFile, sep = "\t" )
rasGeneSets <- list( og = subset( rasPathwayTab, OG.TSG == "OG" )$Gene,
                    tsg = subset( rasPathwayTab, OG.TSG == "TSG" )$Gene )
rasGeneSets <- lapply( rasGeneSets, as.character )
ras_pathway_genes <- unlist( rasGeneSets )

rnaseq_se_file <- "rnaseq_se.rds"
rnaseq_se <- readRDS( file.path( "data", "CCLE_rnaseq_se.rds" )
```

```{r render_mutation_counts}
coldat_ind <- intersect( colnames( colData( rnaseq_se ) ),
                        ras_pathway_genes )
mut_bar_gg <- colData( rnaseq_se ) %>%
    as.data.frame( ) %>%
    dplyr::select( coldat_ind, Tumor_Sample_Barcode ) %>%
    gather( genes, call, -Tumor_Sample_Barcode ) %>%
    filter( !is.na( call ) ) %>%
    dplyr::select( -call ) %>%
    distinct( ) %>%
    ggplot( aes( x = genes ) ) +
    geom_bar( ) +
    labs( x = "RAS pathway genes", y = "Driver mutation count" )
print( mut_bar_gg )
```

# RAS84 Clustering

We clustered the LUNG CCLE cell lines into two and five groups using the
RAS84 genes.

```{r clustering}
ras84_f <- rowData( rnaseq_se )$ensembl_gene_id %in% ras84_map$CCLE_RNASeq
lung_ras84_se <- rnaseq_se[ ras84_f, rnaseq_se$cancer_type == "LUNG" ]

cluster_no <- 2
vst_mat <- assays( lung_ras84_se )$vst
hc <- hclust( dist( t( vst_mat ) ), method = "ward.D2" )
hcd <- as.dendrogram( hc )
ct <- cutree( hc, k = cluster_no )
ct_df <- ct %>%
    as.data.frame( ) %>%
    dplyr::rename( ct_cluster = 1 ) %>%
    rownames_to_column( var = "Tumor_Sample_Barcode" )
ct_means <- vst_mat %>%
    as.data.frame( ) %>%
    tidyr::gather( Tumor_Sample_Barcode, value ) %>%
    left_join( ct_df, by = "Tumor_Sample_Barcode" ) %>%
    group_by( ct_cluster ) %>%
    dplyr::summarize( mean = mean( value ) ) %>%
    arrange( mean ) %>%
    mutate( RSG_2 = paste( "RSG", 0:( cluster_no - 1 ), sep = "_" ) )
RAS84_class <- ct_df %>%
    left_join( ct_means, by = "ct_cluster" ) %>%
    mutate( RSG_2 = factor( RSG_2 ) ) %>%
    dplyr::select( Tumor_Sample_Barcode, RSG_2 )

colData( lung_ras84_se ) <- colData( lung_ras84_se ) %>%
    as.data.frame( ) %>%
    left_join( RAS84_class, by = "Tumor_Sample_Barcode", suffix = c( "", "" ) ) %>%
    DataFrame( row.names = .$Tumor_Sample_Barcode )

colors <- c( "red", "blue" )
hcd %>%  set( "leaves_pch", 19 ) %>%
  set( "leaves_cex", 2 ) %>%
    set( "leaves_col", colors[ ct ][ hc$order ] ) %>%
    set( "labels", "" ) %>%
    plot( )

pca <- prcomp( t( assays( lung_ras84_se )$vst ) )
pca_gg <- pca$x[, 1:2 ] %>%
    as.data.frame( ) %>%
    rownames_to_column( var = "Tumor_Sample_Barcode" ) %>%
    left_join( as.data.frame( colData( lung_ras84_se )[, c( "Tumor_Sample_Barcode", "RSG_2" ) ] ),
              by = "Tumor_Sample_Barcode" ) %>%
    ggplot( aes( x = PC1, y = PC2, color = RSG_2 ) ) +
    geom_point( )
print( pca_gg )
```

## Visualisation - Heatmaps

A Heatmap showing the expression of the RAS84 signature across the two clusters. Patient KRAS mutation status is shown for each sample row indicated in purple.

### Sample mutation annotation

```{r mut_anno}
anno_row_df <- colData( lung_ras84_se ) %>%
    as.data.frame( ) %>%
    mutate( kras_mut = !is.na( KRAS ) ) %>%
    dplyr::select( kras_mut, Tumor_Sample_Barcode ) %>%
    column_to_rownames( var = "Tumor_Sample_Barcode" )

cols_row <- list( kras_mut = c( 'TRUE' = my_purple, 'FALSE' = "white" ) )
hm_anno_row <- rowAnnotation( df = anno_row_df,
                             col = cols_row,
                             na_col = "white",
                             gap = unit( 1, "mm" ),
                             show_legend = FALSE,
                             show_annotation_name = TRUE,
                             annotation_name_side = "top",
                             annotation_legend_param = list( title_gp = gpar( fontsize = 12 ),
                                                            labels_gp = gpar( fontsize = 10 ) ),
                             simple_anno_size = unit( 3, "mm" ) )
```

### Heatmap

```{r render_heatmap}
hmcols <- colorpanel( 1000, my_blue, "white", my_red )

hmdat <- assays( lung_ras84_se )$vst %>%
                               as.data.frame( ) %>%
                               rownames_to_column( var = "gene_id" ) %>%
                               left_join( as.data.frame( rowData( lung_ras84_se )[, c( "gene_id", "external_gene_name" ) ] ), by = "gene_id" ) %>%
                               dplyr::select( -gene_id ) %>%
                               gather( Tumor_Sample_Barcode, value, -external_gene_name ) %>%
                               spread( external_gene_name, value ) %>%
                               left_join( as.data.frame( colData( lung_ras84_se )[, c( "Tumor_Sample_Barcode", "RSG_2", "RSG_5" ) ] ), by = "Tumor_Sample_Barcode" ) %>%
                               column_to_rownames( var = "Tumor_Sample_Barcode" )

hmdat <- hmdat[ rownames( anno_row_df ), ]
hmmat <- hmdat %>%
    dplyr::select( -RSG_2, -RSG_5 ) %>%
    as.matrix( ) %>%
    t( )

hmmat <- hmmat - rowMedians( hmmat )
hm <- Heatmap( t( hmmat ),
              name = "Expression", 
              clustering_method_rows = "ward.D2",
              clustering_method_columns = "ward.D2",
              show_column_names = TRUE,
              show_row_names = FALSE,
              col = colorRamp2(
                  seq( from = -2, to = 2, length.out = length( hmcols ) ),
                  hmcols ),
              row_split = hmdat$RSG_2,
              column_title = "RA84",
              column_title_gp = gpar( fontsize = 14 ),
              column_names_gp = gpar( fontsize = 4 ),
              heatmap_legend_param = list( title_gp = gpar( fontsize = 14 ),
                                          labels_gp = gpar( fontsize = 4 ) ),
              show_heatmap_legend = FALSE,
              cluster_row_slices = FALSE,
              row_names_max_width = unit( 10, "mm" ) )
draw( hm + hm_anno_row )

png( file = file.path( adat$plot.path, "hm.png" ) )
print( draw( hm + hm_anno_row ) )
dev.off( )
```

# RNA-Seq drug response intregration {.tabset .tabset-fade}

We integrated the CCLE RNA-Seq data with the drug sensitivity
data. This is done via the CELL_LINE_ID label in each SE object.

```{r data_integration}
## CCLE data
CCLE_coldat <- colData( lung_ras84_se ) %>%
    as.data.frame( ) %>%
    dplyr::select( Tumor_Sample_Barcode, CELL_LINE_ID, RSG_2, RSG_5, KRASmut, RAS_pathway )

## drug data
drug_ras84_df_l <- map( drug_se_l, function( se ) assays( se )$response ) %>%
    map( as.data.frame ) %>%
    map( rownames_to_column, var = "DRUG_NAME" ) %>%
    map( gather, CELL_LINE_NAME, value, -DRUG_NAME )

## integrate
drug_ras84_df_l <- map2( drug_ras84_df_l, names( drug_ras84_df_l ), function( df, data_n ) {
    df %>%
        left_join( as.data.frame( colData( drug_se_l[[ data_n ]] )[ ,c( "CELL_LINE_NAME", "CELL_LINE_ID" ) ] ), by = "CELL_LINE_NAME" ) %>%
        inner_join( CCLE_coldat, by = "CELL_LINE_ID" ) %>%
        left_join( as.data.frame( rowData( drug_se_l[[ data_n ]] ) ), by = "DRUG_NAME" )
} )

drug_ras84_df_l$GDSC1 <- drug_ras84_df_l$GDSC1 %>%
    filter( !DRUG_NAME %in% c( "Bleomycin (50 uM)", "Bleomycin (10 uM)" ) )

drug_bar_gg <- map( drug_ras84_df_l, pull, RSG_2 ) %>%
    melt( value.name = "RSG" ) %>%
    rename( dataset = L1 ) %>%
    ggplot( aes( x = dataset, fill = RSG ) ) +
    geom_bar( position = position_dodge() )
print( drug_bar_gg )

upset_l <- map( drug_ras84_df_l, function( df ) {
    list( RSG_1 = df %>%
              filter( RSG_2 == "RSG_1" ) %>%
              dplyr::select( CELL_LINE_ID ) %>%
              distinct( ) %>%
              pull( CELL_LINE_ID ),
         RSG_0 = df %>%
             filter( RSG_2 == "RSG_0" ) %>%
             dplyr::select( CELL_LINE_ID ) %>%
             distinct( ) %>%
             pull( CELL_LINE_ID ),
         KRASmut = df %>%
             filter( KRASmut == "mutant" ) %>%
             dplyr::select( CELL_LINE_ID ) %>%
             distinct( ) %>%
             pull( CELL_LINE_ID ),
         KRASwt = df %>%
             filter( KRASmut == "wt" ) %>%
             dplyr::select( CELL_LINE_ID ) %>%
             distinct( ) %>%
             pull( CELL_LINE_ID ) )
} )
```
```{r render_upset,results='asis',echo=FALSE}
for( i in names( upset_l ) ) {
    cat( "## ", i, " \n" )
    print( upset( fromList( upset_l[[ i ]] ), order.by = "freq" ) )
    cat( " \n\n" )
}
```

# Drug Screen

## KRAS mutation

### GDSC1 data

We ran a differential drug sensitivity analysis using the
KRAS mutation data as provided by GDSC1. This analysis was run on all lung
cell lines with KRAS mutation genotype calls provided by GDSC1. The numbers per
group are shown below.

The volcano plot and results table below are the same as the results
presented on the [GDSC website](https://www.cancerrxgene.org/feature/KRAS_mut/152/volcano?screening_set=GDSC1&tissue=LUAD).

It is unclear which KRAS mutations were selected. The data simply
labels the cell lines as mutated or not. Below we cross reference the
mutated cell lines with the CCLE to determine the genotype of the reported
KRAS mutants.

```{r krasmut_gdsc1}
krasmut_n_gdsc1_gg <- colData( gdsc_se_l$GDSC1 ) %>%
    as.data.frame( ) %>%
    filter( !is.na( is_mutated ) ) %>%
    mutate( krasmut_label = ifelse( is_mutated == 1, "mutated", "wt" ) ) %>%
    ggplot( aes( x = krasmut_label ) ) +
    geom_bar( ) +
    theme( axis.text.x = element_text( angle = 90, vjust = 0.5, hjust=1 ) ) +
    labs( x = "KRAS mutation status" )

drug_gdsc1_mat <- assays( gdsc_se_l$GDSC1 )$response
drug_gdsc1_df <- drug_gdsc1_mat %>%
    as.data.frame( ) %>%
    rownames_to_column( var = "DRUG_NAME" ) %>%
    gather( CELL_LINE_NAME, value, -DRUG_NAME ) %>%
    left_join( as.data.frame( colData( gdsc_se_l$GDSC1 ) ), by = "CELL_LINE_NAME" ) %>%
    filter( !is.na( value ) ) %>%
    filter( !is.na( is_mutated ) ) %>%
    mutate( krasmut_label = ifelse( is_mutated == 1, "mutated", "wt" ) ) %>%
    mutate( krasmut_label = factor( krasmut_label, levels = c( "wt", "mutated" ) ) )

drugs_fit_krasmut_gdsc1 <- drug_gdsc1_df %>%
    nest( data = -DRUG_NAME ) %>% 
    mutate(
        fit = map( data, ~ lm( value ~ krasmut_label, data = .x ) ),
        tidied = map( fit, tidy )
        ) %>% 
    unnest( tidied )

drugs_res_krasmut_gdsc1 <- drugs_fit_krasmut_gdsc1 %>%
    filter( term == "krasmut_labelmutated" ) %>%
    mutate( padj = p.adjust( p.value, method = "fdr" ) ) %>%
    arrange( padj ) %>%
    mutate( sensitivity_label = ifelse( estimate < 0 & p.value < 0.001, "sensitive", "no effect" ) ) %>%
    mutate( sensitivity_label = factor( sensitivity_label,
                                        levels = c( "sensitive", "no effect" ) ) )

krasmut_gdsc1_volc_gg <- drugs_res_krasmut_gdsc1 %>%
    ggplot( aes( x = estimate, y = -log10( p.value ), color = sensitivity_label ) ) +
    geom_point( ) +
    geom_text_repel( data = filter( drugs_res_krasmut_gdsc1, sensitivity_label == "sensitive" ),
                    aes( x = estimate, y = -log10( p.value ), label = DRUG_NAME ),
                    inherit.aes = FALSE ) +
    scale_color_manual( values = c( "green", "grey" ) )
print( krasmut_gdsc1_volc_gg )
drugs_res_krasmut_gdsc1 %>%
    dplyr::select( DRUG_NAME, estimate, p.value, padj ) %>%
    DT::datatable( )
```

### CCLE & drug sensitivity data intersection

We compared KRAS mutant to wildtype cell lines using the CCLE KRAS
genotype calls and integrated drug sensitivity data. It is noted that
we prefiltered the CCLE reported KRAS mutants for known driver
genotypes detailed above.

We compared the CCLE and GDSC1 KRAS mutation labels. As shown,
there are 13 KRAS mutant GDSC1 cell lines called as WT in the
CCLE. These are probably KRAS mutants that are not known drivers as
defined above.

```{r ccle_gdsc_kras_mutation_comparison}
gdsc1_coldat <- colData( drug_se_l[[1]] ) %>%
    as.data.frame( ) %>%
    dplyr::select( CELL_LINE_ID, is_mutated ) %>%
    rename( gdsc1_krasmut = is_mutated ) %>%
    mutate( gdsc1_krasmut = ifelse( is.na( gdsc1_krasmut ), "unknown",
                            ifelse( gdsc1_krasmut == 1, "mutant", "wt" ) ) ) %>%
    mutate( gdsc1_krasmut = factor( gdsc1_krasmut, levels = c( "wt", "mutant", "unknown" ) ) )

ccle_gdsc1_krasmut_df <- colData( lung_ras84_se ) %>%
    as.data.frame( ) %>%
    dplyr::select( CELL_LINE_ID, KRASmut_driver, KRAS ) %>%
    rename( ccle_krasmut = KRASmut_driver ) %>%
    rename( kras_genotype = KRAS ) %>%
    inner_join( gdsc1_coldat, by = "CELL_LINE_ID" )

table( ccle_gdsc1_krasmut_df[, c( "ccle_krasmut", "gdsc1_krasmut" ) ] )
```
```{r krasmut_ccle}
drugs_filtered_krasmut_l <- map( drug_ras84_df_l, filter, !is.na( value ) ) %>%
    map( count, DRUG_NAME, KRASmut, .drop = FALSE ) %>%
    map( group_by, DRUG_NAME ) %>%
    map( summarize, drug_filter = all( n > 0 ) ) %>%
    map( filter, drug_filter )

drugs_fit_krasmut_l <- map( drug_ras84_df_l, filter, !is.na( value ) ) %>%
    map2( names( drug_ras84_df_l ), function( df, n ) {
    filter( df, DRUG_NAME %in% drugs_filtered_krasmut_l[[ n ]]$DRUG_NAME )
} ) %>%    
    map( nest, data = -DRUG_NAME ) %>% 
    map( mutate,
        fit = map( data, ~ lm( value ~ KRASmut, data = .x ) ),
        tidied = map( fit, tidy )
        ) %>% 
    map( unnest, tidied )

drugs_res_krasmut_l <- map( drugs_fit_krasmut_l, filter, term == "KRASmutmutant" ) %>%
    map( mutate, padj = p.adjust( p.value, method = "fdr" ) ) %>%
    map( arrange, padj ) %>%
    map( mutate, sensitivity_label = ifelse( estimate > 0, "resistant", "sensitive" ) )

drugs_res_krasmut_l <- map2( drugs_res_krasmut_l, names( drugs_res_krasmut_l ), function( df, data_n ) {
    left_join( df, drugs_filtered_krasmut_l[[ data_n ]], by = "DRUG_NAME", suffix = c( "", "" ) ) %>%
        left_join( as.data.frame( rowData( drug_se_l[[ data_n ]] )[, c( "DRUG_NAME", "TARGET_CATEGORY" ) ] ),
                  by = "DRUG_NAME", suffix = c( "", "" ) )
} )

drugs_decide_krasmut_l <- map( drugs_res_krasmut_l, filter, padj < 0.05 ) %>%
    map( filter, abs( estimate ) > 1 ) %>%
    map( function( df ) split( df, df$sensitivity_label ) )

drugs_phyper_krasmut_res <- map2( drugs_decide_krasmut_l, names( drugs_decide_krasmut_l ), function( l, data_n ) {
    map( l, function( df ) {
        hit_pathways <- table( df$TARGET_CATEGORY )
        map( names( hit_pathways ), function( pathway_name ) {
            q <- hit_pathways[ pathway_name ] ## white balls extracted
            m <- sum( drugs_res_krasmut_l[[ data_n ]]$TARGET_CATEGORY == pathway_name ) ## white ball total
            n <- nrow( drugs_res_krasmut_l[[ data_n ]] ) - m ## black ball total
            p <- nrow( df ) ## balls extracted
            pval <- phyper( q, m, n, p, lower.tail = FALSE )
            data.frame( fg_pathway_n_selected = q,
                       fg_pathway_n = m,
                       bg_n = n,
                       sig_list_n = p,
                       p.value = pval,
                       pathway = pathway_name )
        } ) %>%
            bind_rows( ) %>%
            mutate( padj = p.adjust( p.value, method = "fdr" ) ) %>%
            arrange( padj ) %>%
            filter( padj < 0.05 ) %>%
            filter( fg_pathway_n_selected > 1 ) %>%
            filter( pathway != "Unclassified" )
    } )
} )

drug_pathway_krasmut <- map( drugs_phyper_krasmut_res[ 1:2 ], bind_rows, .id = "sensitivity_label" ) %>%
    bind_rows( .id = "dataset" ) %>%
    droplevels( ) %>%
    unite( "key", dataset, sensitivity_label, pathway, remove = FALSE )
drug_pathway_krasmut_l <- split( drug_pathway_krasmut, drug_pathway_krasmut$pathway )
```

```{r krasmut_vol}
gdsc_res_krasmut_df <- drugs_res_krasmut_l[ 1:2 ] %>%
    bind_rows( .id = "dataset" ) %>%
    mutate( signif = padj < 0.05 & abs( estimate ) > 1 ) %>%
    unite( "key", dataset, sensitivity_label, TARGET_CATEGORY, remove = FALSE )

plotdat_gdsc_krasmut <- map( drug_pathway_krasmut_l, function( df ) {
    gdsc_res_krasmut_df %>%
        mutate( highlight = key %in% df$key & signif ) %>%
        mutate( facet = df$pathway[ 1 ] )
} ) %>%
    bind_rows( .id = "target_category" )

vol_gdsc_krasmut_gg <- ggplot( gdsc_res_krasmut_df, aes( x = estimate, y = -log10( p.value ) ) ) +
    geom_point( color = "#EEEEEE", alpha = 0.5 ) +
    geom_point( data = filter( plotdat_gdsc_krasmut, highlight ),
               aes( color = TARGET_CATEGORY ), size = 3 ) +
    geom_vline( xintercept = 0 ) +
    geom_vline( xintercept = c( -1, 1 ), linetype = "dashed" ) +
    geom_text_repel( data = filter( plotdat_gdsc_krasmut, highlight & sensitivity_label == "resistant" ),
                    aes( x = estimate, y = -log10( p.value ), colour = TARGET_CATEGORY, label = DRUG_NAME ),
                    inherit.aes = FALSE,
                    xlim = c( 3, NA ),
                    nudge_x      = 0.15,
                    direction    = "y",
                    hjust        = -.5,
                    segment.size = 0.2,
                    point.padding = 1,
                    size = 4 ) +
    geom_text_repel( data = filter( plotdat_gdsc_krasmut, highlight & sensitivity_label == "sensitive" ),
                    aes( x = estimate, y = -log10( p.value ), colour = TARGET_CATEGORY, label = DRUG_NAME ),
                    inherit.aes = FALSE,
                    xlim = c( NA, -2 ),
                    nudge_x      = 0.15,
                    direction    = "y",
                    hjust        = .5,
                    segment.size = 0.2,
                    point.padding = 1,
                    size = 4 ) +
    xlim( c( -4, 4 ) ) +
    labs( title = "KRAS mutated" ) +
    plot_formatter( ) +
    theme( legend.position = "none",
           panel.grid.major = element_blank( ),
           panel.grid.minor = element_blank( ),
           strip.background = element_blank( ),
           text = element_text( size = 16 ) ) +
    labs( x = "effect size (delta ln IC50)" ) +
    xlim( c( -4, 4 ) ) +
    ylim( c( 0, 12 ) )

print( vol_gdsc_krasmut_gg )

cairo_pdf( file = file.path( adat$plot.path, "vol_gdsc_krasmut_wrap.pdf" ),
           width = 4, height = 4)
print( vol_gdsc_krasmut_gg )
dev.off( )
```

## RAS pathway mutated

We compared IC50 values across RAS pathway mutated and RAS pathway
wild type cell line groups. We used RRAS pathway genes as defined in
[Oncogenic Signalling Pathways in The Cancer Genome Atlas](https://www.sciencedirect.com/science/article/pii/S0092867418303593?via%3Dihub)
and cancer driver mutations reported in [Comprehensive
Characterization of Cancer Driver Genes and
Mutations](https://www.sciencedirect.com/science/article/pii/S009286741830237X?via%3Dihub).

```{r ras_pathway}
drugs_filtered_ras_pathway_l <- map( drug_ras84_df_l, filter, !is.na( value ) ) %>%
    map( count, DRUG_NAME, RAS_pathway, .drop = FALSE ) %>%
    map( group_by, DRUG_NAME ) %>%
    map( summarize, drug_filter = all( n > 0 ) ) %>%
    map( filter, drug_filter )

drugs_fit_ras_pathway_l <- map( drug_ras84_df_l, filter, !is.na( value ) ) %>%
    map2( names( drug_ras84_df_l ), function( df, n ) {
    filter( df, DRUG_NAME %in% drugs_filtered_ras_pathway_l[[ n ]]$DRUG_NAME )
} ) %>%    
    map( nest, data = -DRUG_NAME ) %>% 
    map( mutate,
        fit = map( data, ~ lm( value ~ RAS_pathway, data = .x ) ),
        tidied = map( fit, tidy )
        ) %>% 
    map( unnest, tidied )

drugs_res_ras_pathway_l <- map( drugs_fit_ras_pathway_l, filter, term == "RAS_pathwayTRUE" ) %>%
    map( mutate, padj = p.adjust( p.value, method = "fdr" ) ) %>%
    map( arrange, padj ) %>%
    map( mutate, sensitivity_label = ifelse( estimate > 0, "resistant", "sensitive" ) )

drugs_res_ras_pathway_l <- map2( drugs_res_ras_pathway_l, names( drugs_res_ras_pathway_l ), function( df, data_n ) {
    left_join( df, drugs_filtered_ras_pathway_l[[ data_n ]], by = "DRUG_NAME", suffix = c( "", "" ) ) %>%
        left_join( as.data.frame( rowData( drug_se_l[[ data_n ]] )[, c( "DRUG_NAME", "TARGET_CATEGORY" ) ] ),
                  by = "DRUG_NAME", suffix = c( "", "" ) )
} )

drugs_decide_ras_pathway_l <- map( drugs_res_ras_pathway_l, filter, padj < 0.05 ) %>%
    map( filter, abs( estimate ) > 0 ) %>%
    map( function( df ) split( df, df$sensitivity_label ) )

drugs_phyper_ras_pathway_res <- map2( drugs_decide_ras_pathway_l, names( drugs_decide_ras_pathway_l ), function( l, data_n ) {
    map( l, function( df ) {
        hit_pathways <- table( df$TARGET_CATEGORY )
        map( names( hit_pathways ), function( pathway_name ) {
            q <- hit_pathways[ pathway_name ] ## white balls extracted
            m <- sum( drugs_res_ras_pathway_l[[ data_n ]]$TARGET_CATEGORY == pathway_name ) ## white ball total
            n <- nrow( drugs_res_ras_pathway_l[[ data_n ]] ) - m ## black ball total
            p <- nrow( df ) ## balls extracted
            pval <- phyper( q, m, n, p, lower.tail = FALSE )
            data.frame( fg_pathway_n_selected = q,
                       fg_pathway_n = m,
                       bg_n = n,
                       sig_list_n = p,
                       p.value = pval,
                       pathway = pathway_name )
        } ) %>%
            bind_rows( ) %>%
            mutate( padj = p.adjust( p.value, method = "fdr" ) ) %>%
            arrange( padj ) %>%
            filter( padj < 0.05 ) %>%
            filter( fg_pathway_n_selected > 1 ) %>%
            filter( pathway != "Unclassified" )
    } )
} )

drug_pathway_ras_pathway <- map( drugs_phyper_ras_pathway_res[ 1:2 ], bind_rows, .id = "sensitivity_label" ) %>%
    bind_rows( .id = "dataset" ) %>%
    droplevels( ) %>%
    unite( "key", dataset, sensitivity_label, pathway, remove = FALSE )
drug_pathway_ras_pathway_l <- split( drug_pathway_ras_pathway, drug_pathway_ras_pathway$pathway )

gdsc_res_df <- drugs_res_ras_pathway_l[ 1:2 ] %>%
    bind_rows( .id = "dataset" ) %>%
    mutate( signif = padj < 0.05 & abs( estimate ) > 1 ) %>%
    unite( "key", dataset, sensitivity_label, TARGET_CATEGORY, remove = FALSE )

plotdat <- map( drug_pathway_ras_pathway_l, function( df ) {
    gdsc_res_df %>%
        mutate( highlight = key %in% df$key & signif ) %>%
        mutate( facet = df$pathway[ 1 ] )
} ) %>%
    bind_rows( .id = "target_category" )

vol_gg <- ggplot( plotdat, aes( x = estimate, y = -log10( p.value ) ) ) +
    geom_point( color = "#EEEEEE", alpha = 0.5 ) +
    geom_point( data = filter( plotdat, signif ), color = "black" ) +
    geom_point( data = filter( plotdat, highlight ),
               aes( color = TARGET_CATEGORY ), size = 3 ) +
    geom_vline( xintercept = 0 ) +
    geom_vline( xintercept = c( -1, 1 ), linetype = "dashed" ) +
    geom_text_repel( data = filter( plotdat, highlight & sensitivity_label == "resistant" ),
                    aes( x = estimate, y = -log10( p.value ), colour = TARGET_CATEGORY, label = DRUG_NAME ),
                    inherit.aes = FALSE,
                    xlim = c( 3, NA ),
                    nudge_x      = 0.15,
                    direction    = "y",
                    hjust        = -.5,
                    segment.size = 0.2,
                    point.padding = 1,
                    size = 4 ) +
    geom_text_repel( data = filter( plotdat, highlight & sensitivity_label == "sensitive" ),
                    aes( x = estimate, y = -log10( p.value ), colour = TARGET_CATEGORY, label = DRUG_NAME ),
                    inherit.aes = FALSE,
                    xlim = c( NA, -2 ),
                    nudge_x      = 0.15,
                    direction    = "y",
                    hjust        = .5,
                    segment.size = 0.2,
                    point.padding = 1,
                    size = 4 ) +
    xlim( c( -4, 4 ) ) +
    labs( title = "RAS pathway mutated" ) +
    plot_formatter( ) +
    theme( legend.position = "none",
           panel.grid.major = element_blank( ),
           panel.grid.minor = element_blank( ),
           strip.background = element_blank( ),
           text = element_text( size = 16 ) ) +
    labs( x = "effect size (delta ln IC50)" ) +
    xlim( c( -4, 4 ) ) +
    ylim( c( 0, 12 ) )

print( vol_gg )

cairo_pdf( file = file.path( adat$plot.path, "vol_gdsc_ras_pathway_wrap.pdf" ),
           width = 4, height = 4)
print( vol_gg )
dev.off( )
```

## RAS84 Signature Group

We compare IC50 values across RAS84 high and low cell line groups as
defined by the clustering above. We removed entries with missing IC50
values and drug entries with zero samples in drug, RSG and KRASmut
groups. 

```{r drug_screen_GDSC}
drugs_filtered_l <- map( drug_ras84_df_l, filter, !is.na( value ) ) %>%
    map( count, DRUG_NAME, RSG_2, KRASmut, .drop = FALSE ) %>%
    map( group_by, DRUG_NAME ) %>%
    map( summarize, drug_filter = all( n > 0 ) ) %>%
    map( filter, drug_filter ) 

drugs_fit_l <- map2( drug_ras84_df_l, names( drug_ras84_df_l ), function( df, n ) {
    filter( df, DRUG_NAME %in% drugs_filtered_l[[ n ]]$DRUG_NAME )
} ) %>%    
    map( nest, data = -DRUG_NAME ) %>% 
    map( mutate,
        fit = map( data, ~ lm( value ~ KRASmut + RSG_2, data = .x ) ),
        tidied = map( fit, tidy )
        ) %>% 
    map( unnest, tidied )
```
```{r gdsc1_rgs}
drugs_res_l <- map( drugs_fit_l, filter, term == "RSG_2RSG_1" ) %>%
    map( mutate, padj = p.adjust( p.value, method = "fdr" ) ) %>%
    map( arrange, padj ) %>%
    map( mutate, sensitivity_label = ifelse( estimate > 0, "resistant", "sensitive" ) )

drugs_res_l <- map2( drugs_res_l, names( drugs_res_l ), function( df, data_n ) {
    left_join( df, drugs_filtered_l[[ data_n ]], by = "DRUG_NAME", suffix = c( "", "" ) ) %>%
        left_join( as.data.frame( rowData( drug_se_l[[ data_n ]] )[, c( "DRUG_NAME", "TARGET_CATEGORY" ) ] ),
                  by = "DRUG_NAME", suffix = c( "", "" ) )
} )

drugs_decide_l <- map( drugs_res_l, filter, padj < 0.05 ) %>%
    map( filter, abs( estimate ) > 1 ) %>%
    map( function( df ) split( df, df$sensitivity_label ) )
```
```{r drug_screen_GDSC1_test,eval=FALSE}
test_drug_name <- "Dihydrorotenone"
test_drug_dat <- drug_ras84_df_l$GDSC2 %>%
    filter( DRUG_NAME == test_drug_name )

test_lm <- lm( value ~ KRASmut + RSG_2, data = test_drug_dat )
summary( test_lm )
```

### Enriched drug targets

We identified drug targets that were enriched in drugs showing
statistically significant differential drug sensitivity between the RAS
high and low groups. We used a hypergeometric test to do this.

```{r gdsc1_target_enrich}
drugs_phyper_res <- map2( drugs_decide_l, names( drugs_decide_l ), function( l, data_n ) {
    map( l, function( df ) {
        hit_pathways <- table( df$TARGET_CATEGORY )
        map( names( hit_pathways ), function( pathway_name ) {
            q <- hit_pathways[ pathway_name ] ## white balls extracted
            m <- sum( drugs_res_l[[ data_n ]]$TARGET_CATEGORY == pathway_name ) ## white ball total
            n <- nrow( drugs_res_l[[ data_n ]] ) - m ## black ball total
            p <- nrow( df ) ## balls extracted
            pval <- phyper( q, m, n, p, lower.tail = FALSE )
            data.frame( fg_pathway_n_selected = q,
                       fg_pathway_n = m,
                       bg_n = n,
                       sig_list_n = p,
                       p.value = pval,
                       pathway = pathway_name )
        } ) %>%
            bind_rows( ) %>%
            mutate( padj = p.adjust( p.value, method = "fdr" ) ) %>%
            arrange( padj ) %>%
            filter( fg_pathway_n_selected > 1 ) %>%
            filter( pathway != "Unclassified" )
    } )
} )

drugs_phyper_gdsc_pathways <- drugs_phyper_res
drugs_phyper_res <- map( drugs_phyper_res, function( l ) {
    l %>%
        map( filter, padj < 0.05 )
} )
```

### Visualisation

#### GDSC Enrichment plot

Here we plot the tested drug target pathways with assocaited p-values

```{r gdcs_pathway_enrich_plot}
plotdat <- drugs_phyper_gdsc_pathways[ 1:2 ] %>%
    map( bind_rows, .id = "direction" ) %>%
    bind_rows( ) %>%
    arrange( padj ) %>%
    mutate( signif = padj < 0.05 )

enrich_sens_df <- plotdat %>%
    filter( direction == "sensitive" ) %>%
    arrange( desc( p.value ) ) %>%
    mutate( pathway = factor( pathway, levels = unique( pathway ) ) ) %>%
    mutate( order = as.numeric( pathway ) )
    
enrich_rest_df <- plotdat %>%
    filter( direction == "resistant" ) %>%
    arrange( p.value ) %>%
    mutate( p.value = ifelse( p.value == 0, 0.0001, p.value ) ) %>%
    mutate( pathway = factor( pathway, levels = unique( pathway ) ) ) %>%
    mutate( order = as.numeric( pathway ) + nrow( enrich_sens_df ) )
    
enrich_df <- list( enrich_sens_df, enrich_rest_df ) %>%
    bind_rows( ) %>%
    unite( coloured_by, direction, signif, remove = FALSE ) %>%
    mutate( direction = factor( direction, levels = c( "sensitive", "resistant" ) ) )
    
enrich_gg <- ggplot( enrich_df, aes( x = -log10( p.value ), y = order,
                                    size = fg_pathway_n_selected, colour = coloured_by ) ) +
    geom_point( ) +
    facet_grid( direction ~ 1, scales = "free_y", space = "free_y" ) +
    xlim( c( 0, 4.5 ) ) +
    scale_colour_manual( values = c( my_grey, my_red, my_grey, my_green ) ) +
    scale_y_continuous(
        breaks = enrich_df$order,
        labels = as.character( enrich_df$pathway ) ) +
    plot_formatter( ) 

cairo_pdf( file = file.path( adat$plot.path, "gdsc_enrichment_plot.pdf" ),
          width = 7, height = 3 )
print( enrich_gg   )
dev.off( )

enrich_rest_gg <- plotdat %>%
    filter( direction == "resistant" ) %>%
    mutate( p.value = ifelse( p.value == 0, 0.0001, p.value ) ) %>%
    mutate( pathway = factor( pathway, levels = rev( unique( pathway ) ) ) ) %>%
    ggplot( aes( x = -log10( p.value ), y = pathway,
                 size = fg_pathway_n_selected, colour = signif ) ) +
    geom_point( ) +
    scale_colour_manual( values = c( my_grey, my_red ) ) +
    xlim( c( 0, 5 ) )  +
    plot_formatter( )
cairo_pdf( file = file.path( adat$plot.path, "gdsc_rest_enrichment_plot.pdf" ),
          width = 7, height = 2.5 )
print( enrich_rest_gg  )
dev.off( )
```

#### GDSC volcano plot

```{r gdsc_volc,fig.width=12,fig.height=8}
drug_pathway <- map( drugs_phyper_res[ 1:2 ], bind_rows, .id = "sensitivity_label" ) %>%
    bind_rows( .id = "dataset" ) %>%
    droplevels( ) %>%
    unite( "key", dataset, sensitivity_label, pathway, remove = FALSE )
drug_pathway_l <- split( drug_pathway, drug_pathway$pathway )

gdsc_res_df <- drugs_res_l[ 1:2 ] %>%
    bind_rows( .id = "dataset" ) %>%
    mutate( signif = padj < 0.05 & abs( estimate ) > 1 ) %>%
    unite( "key", dataset, sensitivity_label, TARGET_CATEGORY, remove = FALSE )

plotdat <- map( drug_pathway_l, function( df ) {
    gdsc_res_df %>%
        mutate( highlight = key %in% df$key & signif ) %>%
        mutate( facet = df$pathway[ 1 ] )
} ) %>%
    bind_rows( .id = "target_category" ) %>%
    mutate( facet = factor( facet, levels = c( "ERK MAPK signaling", "DNA replication",
                                               "EGFR signaling", "Mitosis",
                                               "Protein stability and degradation",
                                               "Chromatin histone acetylation") ) )

vol_gg <- ggplot( plotdat, aes( x = estimate, y = -log10( p.value ) ) ) +
    geom_point( color = "#EEEEEE", alpha = 0.5 ) +
    geom_point( data = filter( plotdat, signif ), color = "grey", alpha = 0.5 ) +
    geom_point( data = filter( plotdat, highlight ),
               aes( color = TARGET_CATEGORY ), size = 3, alpha = 0.7 ) +
    geom_vline( xintercept = 0 ) +
    geom_vline( xintercept = c( -1, 1 ), linetype = "dashed" ) +
    geom_text_repel( data = filter( plotdat, highlight & sensitivity_label == "resistant" ),
                    aes( x = estimate, y = -log10( p.value ), colour = TARGET_CATEGORY, label = DRUG_NAME ),
                    inherit.aes = FALSE,
                    xlim = c( 3, NA ),
                    nudge_x      = 0.15,
                    direction    = "y",
                    hjust        = -.5,
                    segment.size = 0.2,
                    point.padding = 1,
                    size = 3 ) +
    geom_text_repel( data = filter( plotdat, highlight & sensitivity_label == "sensitive" ),
                    aes( x = estimate, y = -log10( p.value ), colour = TARGET_CATEGORY, label = DRUG_NAME ),
                    inherit.aes = FALSE,
                    xlim = c( NA, -2 ),
                    nudge_x      = 0.15,
                    direction    = "y",
                    hjust        = .5,
                    segment.size = 0.2,
                    point.padding = 1,
                    size = 3 ) +
    xlim( c( -4, 4 ) ) +
    facet_wrap( ~facet, nrow = 3 ) +
    plot_formatter( ) +
    theme( legend.position = "none",
           panel.grid.major = element_blank( ),
           panel.grid.minor = element_blank( ),
           strip.background = element_blank( ),
           text = element_text( size = 16 ) ) +
    labs( x = "effect size (delta ln IC50)" ) +
    xlim( c( -4, 4 ) ) +
    ylim( c( 0, 12 ) ) +
    labs( title = "RAS84 activity" )
print( vol_gg )

cairo_pdf( file = file.path( adat$plot.path, "vol_gdsc_wrap.pdf" ) )
print( vol_gg )
dev.off( )
```

```{r gdsc_output}
dbase <- map( drugs_decide_l[ 1:2 ], bind_rows ) %>%
    map( dplyr::select,
        DRUG_NAME, estimate, p.value, padj, sensitivity_label, TARGET_CATEGORY ) %>%
    bind_rows( .id = "dataset" ) %>%
    unite( "key", dataset, sensitivity_label, TARGET_CATEGORY, remove = FALSE ) %>%
    mutate( enriched = key %in% drug_pathway$key ) %>%
    dplyr::select( -key )
write.xlsx( dbase, file.path( adat$analysis.path, "RAS84_HighvLow_CCLE_GDSC_sig_compounds.xlsx" ) )
```

### CTRP volcano plots

```{r ctrp_volc,fig.width=12,fig.height=10}
drug_pathway_ctrp <- map( drugs_phyper_res[ 3:4 ], bind_rows, .id = "sensitivity_label" ) %>%
    bind_rows( .id = "dataset" ) %>%
    droplevels( ) %>%
    unite( "key", dataset, sensitivity_label, pathway, remove = FALSE )
drug_pathway_ctrp_l <- split( drug_pathway_ctrp, drug_pathway_ctrp$pathway )

res_ctrp_df <- drugs_res_l[ 3:4 ] %>%
    bind_rows( .id = "dataset" ) %>%
    mutate( signif = padj < 0.05 & abs( estimate ) > 1 ) %>%
    unite( "key", dataset, sensitivity_label, TARGET_CATEGORY, remove = FALSE )

plotdat_ctrp <- map( drug_pathway_ctrp_l, function( df ) {
    res_ctrp_df %>%
        mutate( highlight = key %in% df$key & signif ) %>%
        mutate( facet = df$pathway[ 1 ] )
} ) %>%
    bind_rows( .id = "target_category" )

#    mutate( facet = factor( facet, levels = c( "ERK MAPK signaling", "DNA replication",
#                                               "EGFR signaling", "Mitosis",
#                                              "Protein stability and degradation",
#                                              "Chromatin histone acetylation") ) )

vol_ctrp_gg <- ggplot( plotdat_ctrp, aes( x = estimate, y = -log10( p.value ) ) ) +
    geom_point( color = "#EEEEEE", alpha = 0.5 ) +
    geom_point( data = filter( plotdat_ctrp, signif ), color = "grey", alpha = 0.5 ) +
    geom_point( data = filter( plotdat_ctrp, highlight ),
               aes( color = TARGET_CATEGORY ), size = 3 ) +
    geom_vline( xintercept = 0 ) +
    geom_vline( xintercept = c( -1, 1 ), linetype = "dashed" ) +
    geom_text_repel( data = filter( plotdat_ctrp, highlight & sensitivity_label == "resistant" ),
                    aes( x = estimate, y = -log10( p.value ), colour = TARGET_CATEGORY, label = DRUG_NAME ),
                    inherit.aes = FALSE,
                    xlim = c( 3, NA ),
                    nudge_x      = 0.15,
                    direction    = "y",
                    hjust        = -.5,
                    segment.size = 0.2,
                    point.padding = 1,
                    size = 2 ) +
    geom_text_repel( data = filter( plotdat_ctrp, highlight & sensitivity_label == "sensitive" ),
                    aes( x = estimate, y = -log10( p.value ), colour = TARGET_CATEGORY, label = DRUG_NAME ),
                    inherit.aes = FALSE,
                    xlim = c( NA, -2 ),
                    nudge_x      = 0.15,
                    direction    = "y",
                    hjust        = .5,
                    segment.size = 0.2,
                    point.padding = 1,
                    size = 2 ) +
    xlim( c( -4, 4 ) ) +
    facet_wrap( ~facet, nrow = 4 ) +
    plot_formatter( ) +
    theme( legend.position = "none",
           panel.grid.major = element_blank( ),
           panel.grid.minor = element_blank( ),
          strip.background = element_blank( ) ) +
    labs( x = "effect size (delta AUC)",
          title = "RAS84 activity" )
print( vol_ctrp_gg )
cairo_pdf( file = file.path( adat$plot.path, "vol_ctrp_wrap.pdf" ),
          width = 10, height = 10 )
print( vol_ctrp_gg )
dev.off( )
```
```{r ctrp_output}
dbase <- map( drugs_decide_l[ 3:4 ], bind_rows ) %>%
    map( dplyr::select,
        DRUG_NAME, estimate, p.value, padj, sensitivity_label, TARGET_CATEGORY ) %>%
    bind_rows( .id = "dataset" ) %>%
    unite( "key", dataset, sensitivity_label, TARGET_CATEGORY, remove = FALSE ) %>%
    mutate( enriched = key %in% drug_pathway$key ) %>%
    dplyr::select( -key )
write.xlsx( dbase, file.path( adat$analysis.path, "RAS84_HighvLow_CCLE_CTRP_sig_compounds.xlsx" ) )
```

# The parent signatures

## Clustering

```{r other_sigs_clustering}
sigs_class_l <- map2( signatures, names( signatures ), function( sig, sig_name ) {
    sig_f <- rowData( rnaseq_se )$external_gene_name %in% sig
    lung_sig_se <- rnaseq_se[ sig_f, rnaseq_se$cancer_type == "LUNG" ]
    cluster_no <- 2
    vst_mat <- assays( lung_sig_se )$vst
    hc <- hclust( dist( t( vst_mat ) ), method = "ward.D2" )
    hcd <- as.dendrogram( hc )
    ct <- cutree( hc, k = cluster_no )
    ct_df <- ct %>%
        as.data.frame( ) %>%
        dplyr::rename( ct_cluster = 1 ) %>%
        rownames_to_column( var = "Tumor_Sample_Barcode" )
    ct_means <- vst_mat %>%
        as.data.frame( ) %>%
        tidyr::gather( Tumor_Sample_Barcode, value ) %>%
        left_join( ct_df, by = "Tumor_Sample_Barcode" ) %>%
        group_by( ct_cluster ) %>%
        dplyr::summarize( mean = mean( value ) ) %>%
        arrange( mean ) %>%
        mutate( sig_group = paste( "SG", 0:( cluster_no - 1 ), sep = "_" ) )
    sig_class <- ct_df %>%
        left_join( ct_means, by = "ct_cluster" ) %>%
        mutate( sig_group = factor( sig_group ) ) %>%
        dplyr::select( Tumor_Sample_Barcode, sig_group ) %>%
        rename( !!sig_name := sig_group )
} )
```

## Drug screen

```{r other_sigs_drugscreen}
sig_class_decide_l <- map2( sigs_class_l, names( sigs_class_l ), function( sig_class, sig_n ) {
    drug_ras84_df_l <- map( drug_ras84_df_l, left_join, sig_class,
                           by = "Tumor_Sample_Barcode",
                           suffix = c( "", "" ) )
    form <- as.formula( paste( "value ~ KRASmut +", sig_n ) )
    drugs_fit_l <- map2( drug_ras84_df_l[1:2], names( drug_ras84_df_l[1:2] ), function( df, n ) {
        filter( df, DRUG_NAME %in% drugs_filtered_l[[ n ]]$DRUG_NAME )
    } ) %>%
    map( nest, data = -DRUG_NAME ) %>% 
    map( mutate,
        fit = map( data, ~ lm( form, data = .x ) ),
        tidied = map( fit, tidy )
        ) %>% 
    map( unnest, tidied )
    drugs_res_l <- map( drugs_fit_l, filter, term == paste0( sig_n, "SG_1" ) ) %>%
        map( mutate, padj = p.adjust( p.value, method = "fdr" ) ) %>%
        map( arrange, padj ) %>%
        map( mutate, sensitivity_label = ifelse( estimate > 0, "resistant", "sensitive" ) )
    drugs_res_l <- map2( drugs_res_l, names( drugs_res_l ), function( df, data_n ) {
        left_join( df, drugs_filtered_l[[ data_n ]], by = "DRUG_NAME", suffix = c( "", "" ) ) %>%
            left_join( as.data.frame( rowData( drug_se_l[[ data_n ]] )[, c( "DRUG_NAME", "TARGET_CATEGORY" ) ] ),
                      by = "DRUG_NAME", suffix = c( "", "" ) )
    } )
    drugs_decide_l <- map( drugs_res_l, filter, padj < 0.05 ) %>%
        map( filter, abs( estimate ) > 1 ) %>%
        map( function( df ) split( df, df$sensitivity_label ) )
    drugs_decide_l
} )

drugs_phyper_res_sig_l <- map( sig_class_decide_l, function( drugs_decide_l ) {
    drugs_phyper_res <- map2( drugs_decide_l, names( drugs_decide_l ), function( l, data_n ) {
        map( l, function( df ) {
            hit_pathways <- table( df$TARGET_CATEGORY )
            map( names( hit_pathways ), function( pathway_name ) {
                q <- hit_pathways[ pathway_name ] ## white balls extracted
                m <- sum( drugs_res_l[[ data_n ]]$TARGET_CATEGORY == pathway_name ) ## white ball total
                n <- nrow( drugs_res_l[[ data_n ]] ) - m ## black ball total
                p <- nrow( df ) ## balls extracted
                pval <- phyper( q, m, n, p, lower.tail = FALSE )
                data.frame( fg_pathway_n_selected = q,
                           fg_pathway_n = m,
                           bg_n = n,
                           sig_list_n = p,
                           p.value = pval,
                           pathway = pathway_name )
            } ) %>%
                bind_rows( ) %>%
                mutate( padj = p.adjust( p.value, method = "fdr" ) ) %>%
                arrange( padj ) %>%
                filter( padj < 0.05 ) %>%
                filter( fg_pathway_n_selected > 1 ) %>%
                    filter( pathway != "Unclassified" )
        } )
    } )
} )
```


# Session Info

```{r session}
sessionInfo()
```
